// @universe/models/esm/src/enums/Country.js
import { Country } from '../types/index.js';
const COUNTRIES = {
    [Country.ABW]: {
        name: 'Aruba',
        officialName: 'Aruba',
        alt: [],
        sovereignty: Country.NLD,
        alpha2: 'AW',
        alpha3: Country.ABW,
        code: 533,
    },
    [Country.AFG]: {
        name: 'Afghanistan',
        officialName: 'The Islamic Republic of Afghanistan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AF',
        alpha3: Country.AFG,
        code: 4,
    },
    [Country.AGO]: {
        name: 'Angola',
        officialName: 'The Republic of Angola',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AO',
        alpha3: Country.AGO,
        code: 24,
    },
    [Country.AIA]: {
        name: 'Anguilla',
        officialName: 'Anguilla',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'AI',
        alpha3: Country.AIA,
        code: 660,
    },
    [Country.ALA]: {
        name: 'Åland Islands',
        officialName: 'Åland',
        alt: [],
        sovereignty: Country.FIN,
        alpha2: 'AX',
        alpha3: Country.ALA,
        code: 248,
    },
    [Country.ALB]: {
        name: 'Albania',
        officialName: 'The Republic of Albania',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AL',
        alpha3: Country.ALB,
        code: 8,
    },
    [Country.AND]: {
        name: 'Andorra',
        officialName: 'The Principality of Andorra',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AD',
        alpha3: Country.AND,
        code: 20,
    },
    [Country.ARE]: {
        name: 'United Arab Emirates',
        officialName: 'The United Arab Emirates',
        alt: ['UAE'],
        sovereignty: Country.UN,
        alpha2: 'AE',
        alpha3: Country.ARE,
        code: 784,
    },
    [Country.ARG]: {
        name: 'Argentina',
        officialName: 'The Argentine Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AR',
        alpha3: Country.ARG,
        code: 32,
    },
    [Country.ARM]: {
        name: 'Armenia',
        officialName: 'The Republic of Armenia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AM',
        alpha3: Country.ARM,
        code: 51,
    },
    [Country.ASM]: {
        name: 'American Samoa',
        officialName: 'The Territory of American Samoa',
        alt: ['Samoa'],
        sovereignty: Country.USA,
        alpha2: 'AS',
        alpha3: Country.ASM,
        code: 16,
    },
    [Country.ATA]: {
        name: 'Antarctica',
        officialName: 'All land and ice shelves south of the 60th parallel south',
        alt: [],
        sovereignty: Country.NT,
        alpha2: 'AQ',
        alpha3: Country.ATA,
        code: 10,
    },
    [Country.ATF]: {
        name: 'The French Southern Territories',
        officialName: 'The French Southern and Antarctic Lands',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'TF',
        alpha3: Country.ATF,
        code: 260,
    },
    [Country.ATG]: {
        name: 'Antigua and Barbuda',
        officialName: 'Antigua and Barbuda',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AG',
        alpha3: Country.ATG,
        code: 28,
    },
    [Country.AUS]: {
        name: 'Australia',
        officialName: 'The Commonwealth of Australia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AU',
        alpha3: Country.AUS,
        code: 36,
    },
    [Country.AUT]: {
        name: 'Austria',
        officialName: 'The Republic of Austria',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AT',
        alpha3: Country.AUT,
        code: 40,
    },
    [Country.AZE]: {
        name: 'Azerbaijan',
        officialName: 'The Republic of Azerbaijan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'AZ',
        alpha3: Country.AZE,
        code: 31,
    },
    [Country.BDI]: {
        name: 'Burundi',
        officialName: 'The Republic of Burundi',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BI',
        alpha3: Country.BDI,
        code: 108,
    },
    [Country.BEL]: {
        name: 'Belgium',
        officialName: 'The Kingdom of Belgium',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BE',
        alpha3: Country.BEL,
        code: 56,
    },
    [Country.BEN]: {
        name: 'Benin',
        officialName: 'The Republic of Benin',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BJ',
        alpha3: Country.BEN,
        code: 204,
    },
    [Country.BES]: {
        name: 'Bonaire',
        officialName: 'Bonaire, Sint Eustatius and Saba',
        alt: [],
        sovereignty: Country.NLD,
        alpha2: 'BQ',
        alpha3: Country.BES,
        code: 535,
    },
    [Country.BFA]: {
        name: 'Burkina Faso',
        officialName: 'Burkina Faso',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BF',
        alpha3: Country.BFA,
        code: 854,
    },
    [Country.BGD]: {
        name: 'Bangladesh',
        officialName: "The People's Republic of Bangladesh",
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BD',
        alpha3: Country.BGD,
        code: 50,
    },
    [Country.BGR]: {
        name: 'Bulgaria',
        officialName: 'The Republic of Bulgaria',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BG',
        alpha3: Country.BGR,
        code: 100,
    },
    [Country.BHR]: {
        name: 'Bahrain',
        officialName: 'The Kingdom of Bahrain',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BH',
        alpha3: Country.BHR,
        code: 48,
    },
    [Country.BHS]: {
        name: 'Bahamas',
        officialName: 'The Commonwealth of The Bahamas',
        alt: ['The Bahamas'],
        sovereignty: Country.UN,
        alpha2: 'BS',
        alpha3: Country.BHS,
        code: 44,
    },
    [Country.BIH]: {
        name: 'Bosnia and Herzegovina',
        officialName: 'Bosnia and Herzegovina',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BA',
        alpha3: Country.BIH,
        code: 70,
    },
    [Country.BLM]: {
        name: 'Saint Barthélemy',
        officialName: 'The Collectivity of Saint-Barthélemy',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'BL',
        alpha3: Country.BLM,
        code: 652,
    },
    [Country.BLR]: {
        name: 'Belarus',
        officialName: 'The Republic of Belarus',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BY',
        alpha3: Country.BLR,
        code: 112,
    },
    [Country.BLZ]: {
        name: 'Belize',
        officialName: 'Belize',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BZ',
        alpha3: Country.BLZ,
        code: 84,
    },
    [Country.BMU]: {
        name: 'Bermuda',
        officialName: 'Bermuda',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'BM',
        alpha3: Country.BMU,
        code: 60,
    },
    [Country.BOL]: {
        name: 'Bolivia',
        officialName: 'The Plurinational State of Bolivia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BO',
        alpha3: Country.BOL,
        code: 68,
    },
    [Country.BRA]: {
        name: 'Brazil',
        officialName: 'The Federative Republic of Brazil',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BR',
        alpha3: Country.BRA,
        code: 76,
    },
    [Country.BRB]: {
        name: 'Barbados',
        officialName: 'Barbados',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BB',
        alpha3: Country.BRB,
        code: 52,
    },
    [Country.BRN]: {
        name: 'Brunei Darussalam',
        officialName: 'The Nation of Brunei, the Abode of Peace',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BN',
        alpha3: Country.BRN,
        code: 96,
    },
    [Country.BTN]: {
        name: 'Bhutan',
        officialName: 'The Kingdom of Bhutan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BT',
        alpha3: Country.BTN,
        code: 64,
    },
    [Country.BVT]: {
        name: 'Bouvet Island',
        officialName: 'Bouvet Island',
        alt: [],
        sovereignty: Country.NOR,
        alpha2: 'BV',
        alpha3: Country.BVT,
        code: 74,
    },
    [Country.BWA]: {
        name: 'Botswana',
        officialName: 'The Republic of Botswana',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'BW',
        alpha3: Country.BWA,
        code: 72,
    },
    [Country.CAF]: {
        name: 'The Central African Republic',
        officialName: 'The Central African Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CF',
        alpha3: Country.CAF,
        code: 140,
    },
    [Country.CAN]: {
        name: 'Canada',
        officialName: 'Canada',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CA',
        alpha3: Country.CAN,
        code: 124,
    },
    [Country.CCK]: {
        name: 'The Cocos Islands',
        officialName: 'The Territory of Cocos (Keeling) Islands',
        alt: ['The Keeling Islands'],
        sovereignty: Country.AUS,
        alpha2: 'CC',
        alpha3: Country.CCK,
        code: 166,
    },
    [Country.CHE]: {
        name: 'Switzerland',
        officialName: 'The Swiss Confederation',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CH',
        alpha3: Country.CHE,
        code: 756,
    },
    [Country.CHL]: {
        name: 'Chile',
        officialName: 'The Republic of Chile',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CL',
        alpha3: Country.CHL,
        code: 152,
    },
    [Country.CHN]: {
        name: 'China',
        officialName: "The People's Republic of China",
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CN',
        alpha3: Country.CHN,
        code: 156,
    },
    [Country.CIV]: {
        name: "Cote d'Ivoire",
        officialName: "The Republic of Côte d'Ivoire",
        alt: ['Ivory Coast'],
        sovereignty: Country.UN,
        alpha2: 'CI',
        alpha3: Country.CIV,
        code: 384,
    },
    [Country.CMR]: {
        name: 'Cameroon',
        officialName: 'The Republic of Cameroon',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CM',
        alpha3: Country.CMR,
        code: 120,
    },
    [Country.COD]: {
        name: 'Congo-Kinshasa',
        officialName: 'The Democratic Republic of the Congo',
        alt: ['The Congo'],
        sovereignty: Country.UN,
        alpha2: 'CD',
        alpha3: Country.COD,
        code: 180,
    },
    [Country.COG]: {
        name: 'Congo-Brazzaville',
        officialName: 'The Republic of the Congo',
        alt: ['The Congo Republic', 'West Congo'],
        sovereignty: Country.UN,
        alpha2: 'CG',
        alpha3: Country.COG,
        code: 178,
    },
    [Country.COK]: {
        name: 'The Cook Islands',
        officialName: 'The Cook Islands',
        alt: [],
        sovereignty: Country.NZL,
        alpha2: 'CK',
        alpha3: Country.COK,
        code: 184,
    },
    [Country.COL]: {
        name: 'Colombia',
        officialName: 'The Republic of Colombia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CO',
        alpha3: Country.COL,
        code: 170,
    },
    [Country.COM]: {
        name: 'The Comoros',
        officialName: 'The Union of the Comoros',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KM',
        alpha3: Country.COM,
        code: 174,
    },
    [Country.CPV]: {
        name: 'Cabo Verde',
        officialName: 'The Republic of Cabo Verde',
        alt: ['Cape Verde'],
        sovereignty: Country.UN,
        alpha2: 'CV',
        alpha3: Country.CPV,
        code: 132,
    },
    [Country.CRI]: {
        name: 'Costa Rica',
        officialName: 'The Republic of Costa Rica',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CR',
        alpha3: Country.CRI,
        code: 188,
    },
    [Country.CUB]: {
        name: 'Cuba',
        officialName: 'The Republic of Cuba',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CU',
        alpha3: Country.CUB,
        code: 192,
    },
    [Country.CUW]: {
        name: 'Curaçao',
        officialName: 'The Country of Curaçao',
        alt: [],
        sovereignty: Country.NLD,
        alpha2: 'CW',
        alpha3: Country.CUW,
        code: 531,
    },
    [Country.CXR]: {
        name: 'Christmas Island',
        officialName: 'The Territory of Christmas Island',
        alt: [],
        sovereignty: Country.AUS,
        alpha2: 'CX',
        alpha3: Country.CXR,
        code: 162,
    },
    [Country.CYM]: {
        name: 'The Cayman Islands',
        officialName: 'The Cayman Islands',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'KY',
        alpha3: Country.CYM,
        code: 136,
    },
    [Country.CYP]: {
        name: 'Cyprus',
        officialName: 'The Republic of Cyprus',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'CY',
        alpha3: Country.CYP,
        code: 196,
    },
    [Country.CZE]: {
        name: 'Czech Republic',
        officialName: 'The Czech Republic',
        alt: ['Czechia', 'Czechoslovakia'],
        sovereignty: Country.UN,
        alpha2: 'CZ',
        alpha3: Country.CZE,
        code: 203,
    },
    [Country.DEU]: {
        name: 'Germany',
        officialName: 'The Federal Republic of Germany',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'DE',
        alpha3: Country.DEU,
        code: 276,
    },
    [Country.DJI]: {
        name: 'Djibouti',
        officialName: 'The Republic of Djibouti',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'DJ',
        alpha3: Country.DJI,
        code: 262,
    },
    [Country.DMA]: {
        name: 'Dominica',
        officialName: 'The Commonwealth of Dominica',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'DM',
        alpha3: Country.DMA,
        code: 212,
    },
    [Country.DNK]: {
        name: 'Denmark',
        officialName: 'The Kingdom of Denmark',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'DK',
        alpha3: Country.DNK,
        code: 208,
    },
    [Country.DOM]: {
        name: 'Dominican Republic',
        officialName: 'The Dominican Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'DO',
        alpha3: Country.DOM,
        code: 214,
    },
    [Country.DZA]: {
        name: 'Algeria',
        officialName: "The People's Democratic Republic of Algeria",
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'DZ',
        alpha3: Country.DZA,
        code: 12,
    },
    [Country.ECU]: {
        name: 'Ecuador',
        officialName: 'The Republic of Ecuador',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'EC',
        alpha3: Country.ECU,
        code: 218,
    },
    [Country.EGY]: {
        name: 'Egypt',
        officialName: 'The Arab Republic of Egypt',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'EG',
        alpha3: Country.EGY,
        code: 818,
    },
    [Country.ERI]: {
        name: 'Eritrea',
        officialName: 'The State of Eritrea',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ER',
        alpha3: Country.ERI,
        code: 232,
    },
    [Country.ESH]: {
        name: 'Western Sahara',
        officialName: 'The Sahrawi Arab Democratic Republic',
        alt: [],
        sovereignty: 'disputed',
        alpha2: 'EH',
        alpha3: Country.ESH,
        code: 732,
    },
    [Country.ESP]: {
        name: 'Spain',
        officialName: 'The Kingdom of Spain',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ES',
        alpha3: Country.ESP,
        code: 724,
    },
    [Country.EST]: {
        name: 'Estonia',
        officialName: 'The Republic of Estonia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'EE',
        alpha3: Country.EST,
        code: 233,
    },
    [Country.ETH]: {
        name: 'Ethiopia',
        officialName: 'The Federal Democratic Republic of Ethiopia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ET',
        alpha3: Country.ETH,
        code: 231,
    },
    [Country.FIN]: {
        name: 'Finland',
        officialName: 'The Republic of Finland',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'FI',
        alpha3: Country.FIN,
        code: 246,
    },
    [Country.FJI]: {
        name: 'Fiji',
        officialName: 'The Republic of Fiji',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'FJ',
        alpha3: Country.FJI,
        code: 242,
    },
    [Country.FLK]: {
        name: 'The Falkland Islands',
        officialName: 'The Falkland Islands',
        alt: ['The Malvinas'],
        sovereignty: Country.GBR,
        alpha2: 'FK',
        alpha3: Country.FLK,
        code: 238,
    },
    [Country.FRA]: {
        name: 'France',
        officialName: 'The French Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'FR',
        alpha3: Country.FRA,
        code: 250,
    },
    [Country.FRO]: {
        name: 'The Faroe Islands',
        officialName: 'The Faroe Islands',
        alt: [],
        sovereignty: Country.DNK,
        alpha2: 'FO',
        alpha3: Country.FRO,
        code: 234,
    },
    [Country.FSM]: {
        name: 'Micronesia',
        officialName: 'The Federated States of Micronesia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'FM',
        alpha3: Country.FSM,
        code: 583,
    },
    [Country.GAB]: {
        name: 'Gabon',
        officialName: 'The Gabonese Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GA',
        alpha3: Country.GAB,
        code: 266,
    },
    [Country.GBR]: {
        name: 'United Kingdom',
        officialName: 'The United Kingdom of Great Britain and Northern Ireland',
        alt: ['Wales', 'Scotland', 'England', 'Ireland', 'Northern Ireland', 'United Kingdom', 'The UK', 'UK', 'The United Kingdom'],
        sovereignty: Country.UN,
        alpha2: 'GB',
        alpha3: Country.GBR,
        code: 826,
    },
    [Country.GEO]: {
        name: 'Georgia',
        officialName: 'Georgia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GE',
        alpha3: Country.GEO,
        code: 268,
    },
    [Country.GGY]: {
        name: 'Guernsey',
        officialName: 'The Bailiwick of Guernsey',
        alt: [],
        sovereignty: 'British Crown',
        alpha2: 'GG',
        alpha3: Country.GGY,
        code: 831,
    },
    [Country.GHA]: {
        name: 'Ghana',
        officialName: 'The Republic of Ghana',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GH',
        alpha3: Country.GHA,
        code: 288,
    },
    [Country.GIB]: {
        name: 'Gibraltar',
        officialName: 'Gibraltar',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'GI',
        alpha3: Country.GIB,
        code: 292,
    },
    [Country.GIN]: {
        name: 'Guinea',
        officialName: 'The Republic of Guinea',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GN',
        alpha3: Country.GIN,
        code: 324,
    },
    [Country.GLP]: {
        name: 'Guadeloupe',
        officialName: 'Guadeloupe',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'GP',
        alpha3: Country.GLP,
        code: 312,
    },
    [Country.GMB]: {
        name: 'Gambia',
        officialName: 'The Republic of The Gambia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GM',
        alpha3: Country.GMB,
        code: 270,
    },
    [Country.GNB]: {
        name: 'Guinea-Bissau',
        officialName: 'The Republic of Guinea-Bissau',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GW',
        alpha3: Country.GNB,
        code: 624,
    },
    [Country.GNQ]: {
        name: 'Equatorial Guinea',
        officialName: 'The Republic of Equatorial Guinea',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GQ',
        alpha3: Country.GNQ,
        code: 226,
    },
    [Country.GRC]: {
        name: 'Greece',
        officialName: 'The Hellenic Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GR',
        alpha3: Country.GRC,
        code: 300,
    },
    [Country.GRD]: {
        name: 'Grenada',
        officialName: 'Grenada',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GD',
        alpha3: Country.GRD,
        code: 308,
    },
    [Country.GRL]: {
        name: 'Greenland',
        officialName: 'Kalaallit Nunaat',
        alt: [],
        sovereignty: Country.DNK,
        alpha2: 'GL',
        alpha3: Country.GRL,
        code: 304,
    },
    [Country.GTM]: {
        name: 'Guatemala',
        officialName: 'The Republic of Guatemala',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GT',
        alpha3: Country.GTM,
        code: 320,
    },
    [Country.GUF]: {
        name: 'French Guiana',
        officialName: 'Guyane',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'GF',
        alpha3: Country.GUF,
        code: 254,
    },
    [Country.GUM]: {
        name: 'Guam',
        officialName: 'The Territory of Guam',
        alt: [],
        sovereignty: Country.USA,
        alpha2: 'GU',
        alpha3: Country.GUM,
        code: 316,
    },
    [Country.GUY]: {
        name: 'Guyana',
        officialName: 'The Co-operative Republic of Guyana',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'GY',
        alpha3: Country.GUY,
        code: 328,
    },
    [Country.HKG]: {
        name: 'Hong Kong',
        officialName: 'The Hong Kong Special Administrative Region of China',
        alt: [],
        sovereignty: Country.CHN,
        alpha2: 'HK',
        alpha3: Country.HKG,
        code: 344,
    },
    [Country.HMD]: {
        name: 'Heard Island and McDonald Islands',
        officialName: 'The Territory of Heard Island and McDonald Islands',
        alt: [],
        sovereignty: Country.AUS,
        alpha2: 'HM',
        alpha3: Country.HMD,
        code: 334,
    },
    [Country.HND]: {
        name: 'Honduras',
        officialName: 'The Republic of Honduras',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'HN',
        alpha3: Country.HND,
        code: 340,
    },
    [Country.HRV]: {
        name: 'Croatia',
        officialName: 'The Republic of Croatia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'HR',
        alpha3: Country.HRV,
        code: 191,
    },
    [Country.HTI]: {
        name: 'Haiti',
        officialName: 'The Republic of Haiti',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'HT',
        alpha3: Country.HTI,
        code: 332,
    },
    [Country.HUN]: {
        name: 'Hungary',
        officialName: 'Hungary',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'HU',
        alpha3: Country.HUN,
        code: 348,
    },
    [Country.IDN]: {
        name: 'Indonesia',
        officialName: 'The Republic of Indonesia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ID',
        alpha3: Country.IDN,
        code: 360,
    },
    [Country.IMN]: {
        name: 'Isle of Man',
        officialName: 'The Isle of Man',
        alt: [],
        sovereignty: 'British Crown',
        alpha2: 'IM',
        alpha3: Country.IMN,
        code: 833,
    },
    [Country.IND]: {
        name: 'India',
        officialName: 'The Republic of India',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'IN',
        alpha3: Country.IND,
        code: 356,
    },
    [Country.IOT]: {
        name: 'The British Indian Ocean Territory',
        officialName: 'The British Indian Ocean Territory',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'IO',
        alpha3: Country.IOT,
        code: 86,
    },
    [Country.IRL]: {
        name: 'Ireland',
        officialName: 'Ireland',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'IE',
        alpha3: Country.IRL,
        code: 372,
    },
    [Country.IRN]: {
        name: 'Iran',
        officialName: 'The Islamic Republic of Iran',
        alt: ['Iran', 'Iran Islamic Republic of'],
        sovereignty: Country.UN,
        alpha2: 'IR',
        alpha3: Country.IRN,
        code: 364,
    },
    [Country.IRQ]: {
        name: 'Iraq',
        officialName: 'The Republic of Iraq',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'IQ',
        alpha3: Country.IRQ,
        code: 368,
    },
    [Country.ISL]: {
        name: 'Iceland',
        officialName: 'Iceland',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'IS',
        alpha3: Country.ISL,
        code: 352,
    },
    [Country.ISR]: {
        name: 'Israel',
        officialName: 'The State of Israel',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'IL',
        alpha3: Country.ISR,
        code: 376,
    },
    [Country.ITA]: {
        name: 'Italy',
        officialName: 'The Italian Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'IT',
        alpha3: Country.ITA,
        code: 380,
    },
    [Country.JAM]: {
        name: 'Jamaica',
        officialName: 'Jamaica',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'JM',
        alpha3: Country.JAM,
        code: 388,
    },
    [Country.JEY]: {
        name: 'Jersey',
        officialName: 'The Bailiwick of Jersey',
        alt: [],
        sovereignty: 'British Crown',
        alpha2: 'JE',
        alpha3: Country.JEY,
        code: 832,
    },
    [Country.JOR]: {
        name: 'Jordan',
        officialName: 'The Hashemite Kingdom of Jordan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'JO',
        alpha3: Country.JOR,
        code: 400,
    },
    [Country.JPN]: {
        name: 'Japan',
        officialName: 'Japan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'JP',
        alpha3: Country.JPN,
        code: 392,
    },
    [Country.KAZ]: {
        name: 'Kazakhstan',
        officialName: 'The Republic of Kazakhstan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KZ',
        alpha3: Country.KAZ,
        code: 398,
    },
    [Country.KEN]: {
        name: 'Kenya',
        officialName: 'The Republic of Kenya',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KE',
        alpha3: Country.KEN,
        code: 404,
    },
    [Country.KGZ]: {
        name: 'Kyrgyzstan',
        officialName: 'The Kyrgyz Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KG',
        alpha3: Country.KGZ,
        code: 417,
    },
    [Country.KHM]: {
        name: 'Cambodia',
        officialName: 'The Kingdom of Cambodia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KH',
        alpha3: Country.KHM,
        code: 116,
    },
    [Country.KIR]: {
        name: 'Kiribati',
        officialName: 'The Republic of Kiribati',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KI',
        alpha3: Country.KIR,
        code: 296,
    },
    [Country.KNA]: {
        name: 'Saint Kitts and Nevis',
        officialName: 'Saint Kitts and Nevis',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'KN',
        alpha3: Country.KNA,
        code: 659,
    },
    [Country.KOR]: {
        name: 'Korea',
        officialName: 'The Republic of Korea',
        alt: ['South Korea', 'Korea The Republic of', 'Korea Republic of'],
        sovereignty: Country.UN,
        alpha2: 'KR',
        alpha3: Country.KOR,
        code: 410,
    },
    [Country.KWT]: {
        name: 'Kuwait',
        officialName: 'The State of Kuwait',
        alt: ['Kuwait The State of'],
        sovereignty: Country.UN,
        alpha2: 'KW',
        alpha3: Country.KWT,
        code: 414,
    },
    [Country.LAO]: {
        name: 'Laos',
        officialName: "The Lao People's Democratic Republic",
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LA',
        alpha3: Country.LAO,
        code: 418,
    },
    [Country.LBN]: {
        name: 'Lebanon',
        officialName: 'The Lebanese Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LB',
        alpha3: Country.LBN,
        code: 422,
    },
    [Country.LBR]: {
        name: 'Liberia',
        officialName: 'The Republic of Liberia',
        alt: ['Liberia The Republic of'],
        sovereignty: Country.UN,
        alpha2: 'LR',
        alpha3: Country.LBR,
        code: 430,
    },
    [Country.LBY]: {
        name: 'Libya',
        officialName: 'The State of Libya',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LY',
        alpha3: Country.LBY,
        code: 434,
    },
    [Country.LCA]: {
        name: 'Saint Lucia',
        officialName: 'Saint Lucia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LC',
        alpha3: Country.LCA,
        code: 662,
    },
    [Country.LIE]: {
        name: 'Liechtenstein',
        officialName: 'The Principality of Liechtenstein',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LI',
        alpha3: Country.LIE,
        code: 438,
    },
    [Country.LKA]: {
        name: 'Sri Lanka',
        officialName: 'The Democratic Socialist Republic of Sri Lanka',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LK',
        alpha3: Country.LKA,
        code: 144,
    },
    [Country.LSO]: {
        name: 'Lesotho',
        officialName: 'The Kingdom of Lesotho',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LS',
        alpha3: Country.LSO,
        code: 426,
    },
    [Country.LTU]: {
        name: 'Lithuania',
        officialName: 'The Republic of Lithuania',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LT',
        alpha3: Country.LTU,
        code: 440,
    },
    [Country.LUX]: {
        name: 'Luxembourg',
        officialName: 'The Grand Duchy of Luxembourg',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LU',
        alpha3: Country.LUX,
        code: 442,
    },
    [Country.LVA]: {
        name: 'Latvia',
        officialName: 'The Republic of Latvia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'LV',
        alpha3: Country.LVA,
        code: 428,
    },
    [Country.MAC]: {
        name: 'Macao',
        officialName: 'Macao Special Administrative Region of China',
        alt: [],
        sovereignty: Country.CHN,
        alpha2: 'MO',
        alpha3: Country.MAC,
        code: 446,
    },
    [Country.MAF]: {
        name: 'Saint Martin',
        officialName: 'The Collectivity of Saint-Martin',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'MF',
        alpha3: Country.MAF,
        code: 663,
    },
    [Country.MAR]: {
        name: 'Morocco',
        officialName: 'The Kingdom of Morocco',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MA',
        alpha3: Country.MAR,
        code: 504,
    },
    [Country.MCO]: {
        name: 'Monaco',
        officialName: 'The Principality of Monaco',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MC',
        alpha3: Country.MCO,
        code: 492,
    },
    [Country.MDA]: {
        name: 'Moldova',
        officialName: 'The Republic of Moldova',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MD',
        alpha3: Country.MDA,
        code: 498,
    },
    [Country.MDG]: {
        name: 'Madagascar',
        officialName: 'The Republic of Madagascar',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MG',
        alpha3: Country.MDG,
        code: 450,
    },
    [Country.MDV]: {
        name: 'Maldives',
        officialName: 'The Republic of Maldives',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MV',
        alpha3: Country.MDV,
        code: 462,
    },
    [Country.MEX]: {
        name: 'Mexico',
        officialName: 'The United Mexican States',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MX',
        alpha3: Country.MEX,
        code: 484,
    },
    [Country.MHL]: {
        name: 'The Marshall Islands',
        officialName: 'The Republic of the Marshall Islands',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MH',
        alpha3: Country.MHL,
        code: 584,
    },
    [Country.MKD]: {
        name: 'North Macedonia',
        officialName: 'Republic of North Macedonia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MK',
        alpha3: Country.MKD,
        code: 807,
    },
    [Country.MLI]: {
        name: 'Mali',
        officialName: 'The Republic of Mali',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ML',
        alpha3: Country.MLI,
        code: 466,
    },
    [Country.MLT]: {
        name: 'Malta',
        officialName: 'The Republic of Malta',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MT',
        alpha3: Country.MLT,
        code: 470,
    },
    [Country.MMR]: {
        name: 'Myanmar',
        officialName: 'The Republic of the Union of Myanmar',
        alt: ['Burma'],
        sovereignty: Country.UN,
        alpha2: 'MM',
        alpha3: Country.MMR,
        code: 104,
    },
    [Country.MNE]: {
        name: 'Montenegro',
        officialName: 'Montenegro',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ME',
        alpha3: Country.MNE,
        code: 499,
    },
    [Country.MNG]: {
        name: 'Mongolia',
        officialName: 'The State of Mongolia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MN',
        alpha3: Country.MNG,
        code: 496,
    },
    [Country.MNP]: {
        name: 'The Northern Mariana Islands',
        officialName: 'The Commonwealth of the Northern Mariana Islands',
        alt: ['Mariana Islands', 'Mariana Isl', 'Mariana isle', 'Mariana is', 'Mariana'],
        sovereignty: Country.USA,
        alpha2: 'MP',
        alpha3: Country.MNP,
        code: 580,
    },
    [Country.MOZ]: {
        name: 'Mozambique',
        officialName: 'The Republic of Mozambique',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MZ',
        alpha3: Country.MOZ,
        code: 508,
    },
    [Country.MRT]: {
        name: 'Mauritania',
        officialName: 'The Islamic Republic of Mauritania',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MR',
        alpha3: Country.MRT,
        code: 478,
    },
    [Country.MSR]: {
        name: 'Montserrat',
        officialName: 'Montserrat',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'MS',
        alpha3: Country.MSR,
        code: 500,
    },
    [Country.MTQ]: {
        name: 'Martinique',
        officialName: 'Martinique',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'MQ',
        alpha3: Country.MTQ,
        code: 474,
    },
    [Country.MUS]: {
        name: 'Mauritius',
        officialName: 'The Republic of Mauritius',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MU',
        alpha3: Country.MUS,
        code: 480,
    },
    [Country.MWI]: {
        name: 'Malawi',
        officialName: 'The Republic of Malawi',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MW',
        alpha3: Country.MWI,
        code: 454,
    },
    [Country.MYS]: {
        name: 'Malaysia',
        officialName: 'Malaysia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'MY',
        alpha3: Country.MYS,
        code: 458,
    },
    [Country.MYT]: {
        name: 'Mayotte',
        officialName: 'The Department of Mayotte',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'YT',
        alpha3: Country.MYT,
        code: 175,
    },
    [Country.NAM]: {
        name: 'Namibia',
        officialName: 'The Republic of Namibia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NA',
        alpha3: Country.NAM,
        code: 516,
    },
    [Country.NCL]: {
        name: 'New Caledonia',
        officialName: 'New Caledonia',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'NC',
        alpha3: Country.NCL,
        code: 540,
    },
    [Country.NER]: {
        name: 'Niger',
        officialName: 'The Republic of the Niger',
        alt: ['The Niger'],
        sovereignty: Country.UN,
        alpha2: 'NE',
        alpha3: Country.NER,
        code: 562,
    },
    [Country.NFK]: {
        name: 'Norfolk Island',
        officialName: 'The Territory of Norfolk Island',
        alt: [],
        sovereignty: Country.AUS,
        alpha2: 'NF',
        alpha3: Country.NFK,
        code: 574,
    },
    [Country.NGA]: {
        name: 'Nigeria',
        officialName: 'The Federal Republic of Nigeria',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NG',
        alpha3: Country.NGA,
        code: 566,
    },
    [Country.NIC]: {
        name: 'Nicaragua',
        officialName: 'The Republic of Nicaragua',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NI',
        alpha3: Country.NIC,
        code: 558,
    },
    [Country.NIU]: {
        name: 'Niue',
        officialName: 'Niue',
        alt: [],
        sovereignty: Country.NZL,
        alpha2: 'NU',
        alpha3: Country.NIU,
        code: 570,
    },
    [Country.NLD]: {
        name: 'Netherlands',
        officialName: 'The Kingdom of the Netherlands',
        alt: ['The Netherlands'],
        sovereignty: Country.UN,
        alpha2: 'NL',
        alpha3: Country.NLD,
        code: 528,
    },
    [Country.NOR]: {
        name: 'Norway',
        officialName: 'The Kingdom of Norway',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NO',
        alpha3: Country.NOR,
        code: 578,
    },
    [Country.NPL]: {
        name: 'Nepal',
        officialName: 'The Federal Democratic Republic of Nepal',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NP',
        alpha3: Country.NPL,
        code: 524,
    },
    [Country.NRU]: {
        name: 'Nauru',
        officialName: 'The Republic of Nauru',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NR',
        alpha3: Country.NRU,
        code: 520,
    },
    [Country.NZL]: {
        name: 'New Zealand',
        officialName: 'New Zealand',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NZ',
        alpha3: Country.NZL,
        code: 554,
    },
    [Country.OMN]: {
        name: 'Oman',
        officialName: 'The Sultanate of Oman',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'OM',
        alpha3: Country.OMN,
        code: 512,
    },
    [Country.PAK]: {
        name: 'Pakistan',
        officialName: 'The Islamic Republic of Pakistan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PK',
        alpha3: Country.PAK,
        code: 586,
    },
    [Country.PAN]: {
        name: 'Panama',
        officialName: 'The Republic of Panamá',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PA',
        alpha3: Country.PAN,
        code: 591,
    },
    [Country.PCN]: {
        name: 'Pitcairn',
        officialName: 'The Pitcairn, Henderson, Ducie and Oeno Islands',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'PN',
        alpha3: Country.PCN,
        code: 612,
    },
    [Country.PER]: {
        name: 'Peru',
        officialName: 'The Republic of Perú',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PE',
        alpha3: Country.PER,
        code: 604,
    },
    [Country.PHL]: {
        name: 'Philippines',
        officialName: 'The Republic of the Philippines',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PH',
        alpha3: Country.PHL,
        code: 608,
    },
    [Country.PLW]: {
        name: 'Palau',
        officialName: 'The Republic of Palau',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PW',
        alpha3: Country.PLW,
        code: 585,
    },
    [Country.PNG]: {
        name: 'Papua New Guinea',
        officialName: 'The Independent State of Papua New Guinea',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PG',
        alpha3: Country.PNG,
        code: 598,
    },
    [Country.POL]: {
        name: 'Poland',
        officialName: 'The Republic of Poland',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PL',
        alpha3: Country.POL,
        code: 616,
    },
    [Country.PRI]: {
        name: 'Puerto Rico',
        officialName: 'The Commonwealth of Puerto Rico',
        alt: [],
        sovereignty: Country.USA,
        alpha2: 'PR',
        alpha3: Country.PRI,
        code: 630,
    },
    [Country.PRK]: {
        name: 'North Korea',
        officialName: "The Democratic People's Republic of Korea",
        alt: ['North Korea'],
        sovereignty: Country.UN,
        alpha2: 'KP',
        alpha3: Country.PRK,
        code: 408,
    },
    [Country.PRT]: {
        name: 'Portugal',
        officialName: 'The Portuguese Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PT',
        alpha3: Country.PRT,
        code: 620,
    },
    [Country.PRY]: {
        name: 'Paraguay',
        officialName: 'The Republic of Paraguay',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'PY',
        alpha3: Country.PRY,
        code: 600,
    },
    [Country.PSE]: {
        name: 'Palestine, State of',
        officialName: 'The State of Palestine',
        alt: [],
        sovereignty: 'UN observer',
        alpha2: 'PS',
        alpha3: Country.PSE,
        code: 275,
    },
    [Country.PYF]: {
        name: 'French Polynesia',
        officialName: 'French Polynesia',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'PF',
        alpha3: Country.PYF,
        code: 258,
    },
    [Country.QAT]: {
        name: 'Qatar',
        officialName: 'The State of Qatar',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'QA',
        alpha3: Country.QAT,
        code: 634,
    },
    [Country.REU]: {
        name: 'Réunion',
        officialName: 'Réunion',
        alt: [],
        sovereignty: Country.FRA,
        alpha2: 'RE',
        alpha3: Country.REU,
        code: 638,
    },
    [Country.ROU]: {
        name: 'Romania',
        officialName: 'Romania',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'RO',
        alpha3: Country.ROU,
        code: 642,
    },
    [Country.RUS]: {
        name: 'Russia',
        officialName: 'The Russian Federation',
        alt: ['The Russian Federation', 'Russian Federation'],
        sovereignty: Country.UN,
        alpha2: 'RU',
        alpha3: Country.RUS,
        code: 643,
    },
    [Country.RWA]: {
        name: 'Rwanda',
        officialName: 'The Republic of Rwanda',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'RW',
        alpha3: Country.RWA,
        code: 646,
    },
    [Country.SAU]: {
        name: 'Saudi Arabia',
        officialName: 'The Kingdom of Saudi Arabia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SA',
        alpha3: Country.SAU,
        code: 682,
    },
    [Country.SDN]: {
        name: 'Sudan',
        officialName: 'The Republic of the Sudan',
        alt: ['The Sudan'],
        sovereignty: Country.UN,
        alpha2: 'SD',
        alpha3: Country.SDN,
        code: 729,
    },
    [Country.SEN]: {
        name: 'Senegal',
        officialName: 'The Republic of Senegal',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SN',
        alpha3: Country.SEN,
        code: 686,
    },
    [Country.SGP]: {
        name: 'Singapore',
        officialName: 'The Republic of Singapore',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SG',
        alpha3: Country.SGP,
        code: 702,
    },
    [Country.SGS]: {
        name: 'South Georgia and the South Sandwich Islands',
        officialName: 'South Georgia and the South Sandwich Islands',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'GS',
        alpha3: Country.SGS,
        code: 239,
    },
    [Country.SHN]: {
        name: 'Saint Helena',
        officialName: 'Saint Helena, Ascension and Tristan da Cunha',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'SH',
        alpha3: Country.SHN,
        code: 654,
    },
    [Country.SJM]: {
        name: 'Svalbard',
        officialName: 'Svalbard and Jan Mayen',
        alt: [],
        sovereignty: Country.NOR,
        alpha2: 'SJ',
        alpha3: Country.SJM,
        code: 744,
    },
    [Country.SLB]: {
        name: 'Solomon Islands',
        officialName: 'The Solomon Islands',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SB',
        alpha3: Country.SLB,
        code: 90,
    },
    [Country.SLE]: {
        name: 'Sierra Leone',
        officialName: 'The Republic of Sierra Leone',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SL',
        alpha3: Country.SLE,
        code: 694,
    },
    [Country.SLV]: {
        name: 'El Salvador',
        officialName: 'The Republic of El Salvador',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SV',
        alpha3: Country.SLV,
        code: 222,
    },
    [Country.SMR]: {
        name: 'San Marino',
        officialName: 'The Republic of San Marino',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SM',
        alpha3: Country.SMR,
        code: 674,
    },
    [Country.SOM]: {
        name: 'Somalia',
        officialName: 'The Federal Republic of Somalia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SO',
        alpha3: Country.SOM,
        code: 706,
    },
    [Country.SPM]: {
        name: 'Saint Pierre and Miquelon',
        officialName: 'The Overseas Collectivity of Saint-Pierre and Miquelon',
        alt: ['Saint Pierre', 'Miquelon'],
        sovereignty: Country.FRA,
        alpha2: 'PM',
        alpha3: Country.SPM,
        code: 666,
    },
    [Country.SRB]: {
        name: 'Serbia',
        officialName: 'The Republic of Serbia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'RS',
        alpha3: Country.SRB,
        code: 688,
    },
    [Country.SSD]: {
        name: 'South Sudan',
        officialName: 'The Republic of South Sudan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SS',
        alpha3: Country.SSD,
        code: 728,
    },
    [Country.STP]: {
        name: 'Sao Tome and Principe',
        officialName: 'The Democratic Republic of São Tomé and Príncipe',
        alt: ['São Tomé and Príncipe'],
        sovereignty: Country.UN,
        alpha2: 'ST',
        alpha3: Country.STP,
        code: 678,
    },
    [Country.SUR]: {
        name: 'Suriname',
        officialName: 'The Republic of Suriname',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SR',
        alpha3: Country.SUR,
        code: 740,
    },
    [Country.SVK]: {
        name: 'Slovakia',
        officialName: 'The Slovak Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SK',
        alpha3: Country.SVK,
        code: 703,
    },
    [Country.SVN]: {
        name: 'Slovenia',
        officialName: 'The Republic of Slovenia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SI',
        alpha3: Country.SVN,
        code: 705,
    },
    [Country.SWE]: {
        name: 'Sweden',
        officialName: 'The Kingdom of Sweden',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SE',
        alpha3: Country.SWE,
        code: 752,
    },
    [Country.SWZ]: {
        name: 'Swaziland',
        officialName: 'The Kingdom of Eswatini',
        alt: ['Eswatini'],
        sovereignty: Country.UN,
        alpha2: 'SZ',
        alpha3: Country.SWZ,
        code: 748,
    },
    [Country.SXM]: {
        name: 'Sint Maarten',
        officialName: 'Sint Maarten',
        alt: ['Sint Maarten (Dutch part)'],
        sovereignty: Country.NLD,
        alpha2: 'SX',
        alpha3: Country.SXM,
        code: 534,
    },
    [Country.SYC]: {
        name: 'Seychelles',
        officialName: 'The Republic of Seychelles',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SC',
        alpha3: Country.SYC,
        code: 690,
    },
    [Country.SYR]: {
        name: 'Syria',
        officialName: 'The Syrian Arab Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'SY',
        alpha3: Country.SYR,
        code: 760,
    },
    [Country.TCA]: {
        name: 'Turks and Caicos Islands',
        officialName: 'The Turks and Caicos Islands',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'TC',
        alpha3: Country.TCA,
        code: 796,
    },
    [Country.TCD]: {
        name: 'Chad',
        officialName: 'The Republic of Chad',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TD',
        alpha3: Country.TCD,
        code: 148,
    },
    [Country.TGO]: {
        name: 'Togo',
        officialName: 'The Togolese Republic',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TG',
        alpha3: Country.TGO,
        code: 768,
    },
    [Country.THA]: {
        name: 'Thailand',
        officialName: 'The Kingdom of Thailand',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TH',
        alpha3: Country.THA,
        code: 764,
    },
    [Country.TJK]: {
        name: 'Tajikistan',
        officialName: 'The Republic of Tajikistan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TJ',
        alpha3: Country.TJK,
        code: 762,
    },
    [Country.TKL]: {
        name: 'Tokelau',
        officialName: 'Tokelau',
        alt: [],
        sovereignty: Country.NZL,
        alpha2: 'TK',
        alpha3: Country.TKL,
        code: 772,
    },
    [Country.TKM]: {
        name: 'Turkmenistan',
        officialName: 'Turkmenistan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TM',
        alpha3: Country.TKM,
        code: 795,
    },
    [Country.TLS]: {
        name: 'Timor-Leste',
        officialName: 'The Democratic Republic of Timor-Leste',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TL',
        alpha3: Country.TLS,
        code: 626,
    },
    [Country.TON]: {
        name: 'Tonga',
        officialName: 'The Kingdom of Tonga',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TO',
        alpha3: Country.TON,
        code: 776,
    },
    [Country.TTO]: {
        name: 'Trinidad and Tobago',
        officialName: 'The Republic of Trinidad and Tobago',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TT',
        alpha3: Country.TTO,
        code: 780,
    },
    [Country.TUN]: {
        name: 'Tunisia',
        officialName: 'The Republic of Tunisia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TN',
        alpha3: Country.TUN,
        code: 788,
    },
    [Country.TUR]: {
        name: 'Turkey',
        officialName: 'The Republic of Turkey',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TR',
        alpha3: Country.TUR,
        code: 792,
    },
    [Country.TUV]: {
        name: 'Tuvalu',
        officialName: 'Tuvalu',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'TV',
        alpha3: Country.TUV,
        code: 798,
    },
    [Country.TWN]: {
        name: 'Taiwan',
        officialName: 'The Republic of China',
        alt: [],
        sovereignty: 'disputed',
        alpha2: 'TW',
        alpha3: Country.TWN,
        code: 158,
    },
    [Country.TZA]: {
        name: 'Tanzania, the United Republic of',
        officialName: 'The United Republic of Tanzania',
        alt: ['Tanzania United Republic of'],
        sovereignty: Country.UN,
        alpha2: 'TZ',
        alpha3: Country.TZA,
        code: 834,
    },
    [Country.UGA]: {
        name: 'Uganda',
        officialName: 'The Republic of Uganda',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'UG',
        alpha3: Country.UGA,
        code: 800,
    },
    [Country.UKR]: {
        name: 'Ukraine',
        officialName: 'Ukraine',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'UA',
        alpha3: Country.UKR,
        code: 804,
    },
    [Country.UMI]: {
        name: 'The United States Minor Outlying Islands',
        officialName: 'The United States Minor Outlying Islands',
        alt: [
            'Baker Island',
            'Howland Island',
            'Jarvis Island',
            'Johnston Atoll',
            'Kingman Reef',
            'Midway Atoll',
            'Navassa Island',
            'Palmyra Atoll',
            'Wake Island',
        ],
        sovereignty: Country.USA,
        alpha2: 'UM',
        alpha3: Country.UMI,
        code: 581,
    },
    [Country.URY]: {
        name: 'Uruguay',
        officialName: 'The Oriental Republic of Uruguay',
        alt: ['The Republic of Uruguay'],
        sovereignty: Country.UN,
        alpha2: 'UY',
        alpha3: Country.URY,
        code: 858,
    },
    [Country.USA]: {
        name: 'United States of America',
        officialName: 'The United States of America',
        alt: ['America', 'United States', 'Murica'],
        sovereignty: Country.UN,
        alpha2: 'US',
        alpha3: Country.USA,
        code: 840,
    },
    [Country.UZB]: {
        name: 'Uzbekistan',
        officialName: 'The Republic of Uzbekistan',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'UZ',
        alpha3: Country.UZB,
        code: 860,
    },
    [Country.VAT]: {
        name: 'The Vatican',
        officialName: 'The Holy See',
        alt: ['Vatican City'],
        sovereignty: 'UN observer',
        alpha2: 'VA',
        alpha3: Country.VAT,
        code: 336,
    },
    [Country.VCT]: {
        name: 'Saint Vincent and the Grenadines',
        officialName: 'Saint Vincent and the Grenadines',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'VC',
        alpha3: Country.VCT,
        code: 670,
    },
    [Country.VEN]: {
        name: 'Venezuela',
        officialName: 'The Bolivarian Republic of Venezuela',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'VE',
        alpha3: Country.VEN,
        code: 862,
    },
    [Country.VGB]: {
        name: 'Virgin Islands (British)',
        officialName: 'The Virgin Islands',
        alt: [],
        sovereignty: Country.GBR,
        alpha2: 'VG',
        alpha3: Country.VGB,
        code: 92,
    },
    [Country.VIR]: {
        name: 'Virgin Islands (U.S.)',
        officialName: 'The Virgin Islands of the United States',
        alt: ['Virgin Islands'],
        sovereignty: Country.USA,
        alpha2: 'VI',
        alpha3: Country.VIR,
        code: 850,
    },
    [Country.VNM]: {
        name: 'Vietnam',
        officialName: 'The Socialist Republic of Viet Nam',
        alt: ['Viet Nam'],
        sovereignty: Country.UN,
        alpha2: 'VN',
        alpha3: Country.VNM,
        code: 704,
    },
    [Country.VUT]: {
        name: 'Vanuatu',
        officialName: 'The Republic of Vanuatu',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'VU',
        alpha3: Country.VUT,
        code: 548,
    },
    [Country.WLF]: {
        name: 'Wallis and Futuna',
        officialName: 'The Territory of the Wallis and Futuna Islands',
        alt: ['Wallis', 'Futuna'],
        sovereignty: Country.FRA,
        alpha2: 'WF',
        alpha3: Country.WLF,
        code: 876,
    },
    [Country.WSM]: {
        name: 'The Independent State of Samoa',
        officialName: 'The Independent State of Samoa',
        alt: ['Western Samoa'],
        sovereignty: Country.UN,
        alpha2: 'WS',
        alpha3: Country.WSM,
        code: 882,
    },
    [Country.YEM]: {
        name: 'Yemen',
        officialName: 'The Republic of Yemen',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'YE',
        alpha3: Country.YEM,
        code: 887,
    },
    [Country.ZAF]: {
        name: 'South Africa',
        officialName: 'The Republic of South Africa',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ZA',
        alpha3: Country.ZAF,
        code: 710,
    },
    [Country.ZMB]: {
        name: 'Zambia',
        officialName: 'The Republic of Zambia',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'ZM',
        alpha3: Country.ZMB,
        code: 894,
    },
    [Country.ZWE]: {
        name: 'Zimbabwe',
        officialName: 'The Republic of Zimbabwe',
        alt: ['Rhodesia'],
        sovereignty: Country.UN,
        alpha2: 'ZW',
        alpha3: Country.ZWE,
        code: 716,
    },
    [Country.FOREIGN]: {
        name: 'Foreign',
        officialName: 'Foreign',
        alt: ['Not From the USA'],
        sovereignty: Country.FOREIGN,
        alpha2: '',
        alpha3: Country.FOREIGN,
        code: 999,
    },
    [Country.UN]: {
        name: 'United Nations',
        officialName: 'United Nations',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'UN',
        alpha3: Country.UN,
        code: 0,
    },
    [Country.NT]: {
        name: 'Neutral Territory',
        officialName: 'Neutral Territory',
        alt: [],
        sovereignty: Country.UN,
        alpha2: 'NT',
        alpha3: Country.NT,
        code: -1,
    },
};
export function getCountryDesc(country) {
    const desc = COUNTRIES[country];
    return desc ? { ...desc } : null;
}
const aliases = {};
for (const [key, desc] of Object.entries(COUNTRIES)) {
    aliases[key] = aliases[key] || [];
    aliases[key].push(desc.alpha2);
    aliases[key].push(desc.name);
    for (const alt of [...desc.alt, desc.officialName]) {
        const arr = aliases[key] = aliases[key] || [];
        // Like "Korea, The Republic of"
        arr.push(alt);
        if (alt.includes(' of ')) {
            const inverse = `${alt.split(' of ').reverse().join(', ')} of`;
            arr.push(inverse);
        }
    }
}
export default aliases;
