// @universe/models/esm/src/enums/index.js
import { default as Country } from './Country.js';
import { default as Directional } from './Directional.js';
import { default as FacilityType } from './FacilityType.js';
import { default as LanguageCode } from './LanguageCode.js';
import { default as NameSuffix } from './NameSuffix.js';
import { default as PostNominal } from './PostNominal.js';
import { default as PreNominal } from './PreNominal.js';
import { default as State } from './State.js';
import { default as StreetType } from './StreetType.js';
import { default as UnitAbbr } from './UnitAbbr.js';
import { createAliasesLookup, createLookupFunction } from './utils.js';
export const aliases = {
    Country,
    Directional,
    FacilityType,
    LanguageCode,
    NameSuffix,
    PostNominal,
    PreNominal,
    State,
    StreetType,
    UnitAbbr,
    CountryLookup: createAliasesLookup(Country),
    DirectionalLookup: createAliasesLookup(Directional),
    FacilityTypeLookup: createAliasesLookup(FacilityType),
    LanguageCodeLookup: createAliasesLookup(LanguageCode),
    NameSuffixLookup: createAliasesLookup(NameSuffix),
    PostNominalLookup: createAliasesLookup(PostNominal),
    PreNominalLookup: createAliasesLookup(PreNominal),
    StateLookup: createAliasesLookup(State),
    StreetTypeLookup: createAliasesLookup(StreetType),
    UnitAbbrLookup: createAliasesLookup(UnitAbbr),
};
export const toCountry = createLookupFunction(aliases.CountryLookup);
export const toDirectional = createLookupFunction(aliases.DirectionalLookup);
export const toFacilityType = createLookupFunction(aliases.FacilityTypeLookup);
export const toLanguageCode = createLookupFunction(aliases.LanguageCodeLookup);
export const toPreNominal = createLookupFunction(aliases.PreNominalLookup);
export const toPostNominal = createLookupFunction(aliases.PostNominalLookup);
export const toNameSuffix = createLookupFunction(aliases.NameSuffixLookup);
export const toState = createLookupFunction(aliases.StateLookup);
export const toStreetType = createLookupFunction(aliases.StreetTypeLookup);
export const toUnitAbbr = createLookupFunction(aliases.UnitAbbrLookup);
