// @universe/models/esm/src/enums/State.js
import { Country, State } from '../types/enums.js';
const STATE = {
    // https://en.wikipedia.org/wiki/ISO_3166-2:US
    AL: {
        name: 'Alabama',
        code: State.AL,
        country: Country.USA,
        number: 1,
    },
    AK: {
        name: 'Alaska',
        code: State.AK,
        country: Country.USA,
        number: 2,
    },
    AS: {
        name: 'American Samoa',
        code: State.AS,
        country: Country.USA,
        number: 60,
    },
    AZ: {
        name: 'Arizona',
        code: State.AZ,
        country: Country.USA,
        number: 4,
    },
    AR: {
        name: 'Arkansas',
        code: State.AR,
        country: Country.USA,
        number: 5,
    },
    CA: {
        name: 'California',
        code: State.CA,
        country: Country.USA,
        number: 6,
    },
    CO: {
        name: 'Colorado',
        code: State.CO,
        country: Country.USA,
        number: 8,
    },
    CT: {
        name: 'Connecticut',
        code: State.CT,
        country: Country.USA,
        number: 9,
    },
    DE: {
        name: 'Delaware',
        code: State.DE,
        country: Country.USA,
        number: 10,
    },
    DC: {
        name: 'District of Columbia',
        code: State.DC,
        country: Country.USA,
        number: 11,
    },
    FM: {
        name: 'Federated States of Micronesia',
        code: State.FM,
        country: Country.USA,
        number: 64,
    },
    FL: {
        name: 'Florida',
        code: State.FL,
        country: Country.USA,
        number: 12,
    },
    GA: {
        name: 'Georgia',
        code: State.GA,
        country: Country.USA,
        number: 13,
    },
    GU: {
        name: 'Guam',
        code: State.GU,
        country: Country.USA,
        number: 66,
    },
    HI: {
        name: 'Hawaii',
        code: State.HI,
        country: Country.USA,
        number: 15,
    },
    ID: {
        name: 'Idaho',
        code: State.ID,
        country: Country.USA,
        number: 16,
    },
    IL: {
        name: 'Illinois',
        code: State.IL,
        country: Country.USA,
        number: 17,
    },
    IN: {
        name: 'Indiana',
        code: State.IN,
        country: Country.USA,
        number: 18,
    },
    IA: {
        name: 'Iowa',
        code: State.IA,
        country: Country.USA,
        number: 19,
    },
    KS: {
        name: 'Kansas',
        code: State.KS,
        country: Country.USA,
        number: 20,
    },
    KY: {
        name: 'Kentucky',
        code: State.KY,
        country: Country.USA,
        number: 21,
    },
    LA: {
        name: 'Louisiana',
        code: State.LA,
        country: Country.USA,
        number: 22,
    },
    ME: {
        name: 'Maine',
        code: State.ME,
        country: Country.USA,
        number: 23,
    },
    MH: {
        name: 'Marshall Islands',
        code: State.MH,
        country: Country.USA,
        number: 68,
    },
    MD: {
        name: 'Maryland',
        code: State.MD,
        country: Country.USA,
        number: 24,
    },
    MA: {
        name: 'Massachusetts',
        code: State.MA,
        country: Country.USA,
        number: 25,
    },
    MI: {
        name: 'Michigan',
        code: State.MI,
        country: Country.USA,
        number: 26,
    },
    MN: {
        name: 'Minnesota',
        code: State.MN,
        country: Country.USA,
        number: 27,
    },
    MS: {
        name: 'Mississippi',
        code: State.MS,
        country: Country.USA,
        number: 28,
    },
    MO: {
        name: 'Missouri',
        code: State.MO,
        country: Country.USA,
        number: 29,
    },
    MT: {
        name: 'Montana',
        code: State.MT,
        country: Country.USA,
        number: 30,
    },
    NE: {
        name: 'Nebraska',
        code: State.NE,
        country: Country.USA,
        number: 31,
    },
    NV: {
        name: 'Nevada',
        code: State.NV,
        country: Country.USA,
        number: 32,
    },
    NH: {
        name: 'New Hampshire',
        code: State.NH,
        country: Country.USA,
        number: 33,
    },
    NJ: {
        name: 'New Jersey',
        code: State.NJ,
        country: Country.USA,
        number: 34,
    },
    NM: {
        name: 'New Mexico',
        code: State.NM,
        country: Country.USA,
        number: 35,
    },
    NY: {
        name: 'New York',
        code: State.NY,
        country: Country.USA,
        number: 36,
    },
    NC: {
        name: 'North Carolina',
        code: State.NC,
        country: Country.USA,
        number: 37,
    },
    ND: {
        name: 'North Dakota',
        code: State.ND,
        country: Country.USA,
        number: 38,
    },
    MP: {
        name: 'Northern Mariana Islands',
        code: State.MP,
        country: Country.USA,
        number: 69,
    },
    OH: {
        name: 'Ohio',
        code: State.OH,
        country: Country.USA,
        number: 39,
    },
    OK: {
        name: 'Oklahoma',
        code: State.OK,
        country: Country.USA,
        number: 40,
    },
    OR: {
        name: 'Oregon',
        code: State.OR,
        country: Country.USA,
        number: 41,
    },
    PW: {
        name: 'Palau',
        code: State.PW,
        country: Country.USA,
        number: 70,
    },
    PA: {
        name: 'Pennsylvania',
        code: State.PA,
        country: Country.USA,
        number: 42,
    },
    PR: {
        name: 'Puerto Rico',
        code: State.PR,
        country: Country.USA,
        number: 72,
    },
    RI: {
        name: 'Rhode Island',
        code: State.RI,
        country: Country.USA,
        number: 44,
    },
    SC: {
        name: 'South Carolina',
        code: State.SC,
        country: Country.USA,
        number: 45,
    },
    SD: {
        name: 'South Dakota',
        code: State.SD,
        country: Country.USA,
        number: 46,
    },
    TN: {
        name: 'Tennessee',
        code: State.TN,
        country: Country.USA,
        number: 47,
    },
    TX: {
        name: 'Texas',
        code: State.TX,
        country: Country.USA,
        number: 48,
    },
    UM: {
        name: 'United States Minor Outlying Islands',
        code: State.UM,
        country: Country.USA,
        number: 74,
    },
    UT: {
        name: 'Utah',
        code: State.UT,
        country: Country.USA,
        number: 49,
    },
    VT: {
        name: 'Vermont',
        code: State.VT,
        country: Country.USA,
        number: 50,
    },
    VI: {
        name: 'Virgin Islands',
        code: State.VI,
        country: Country.USA,
        number: 0,
    },
    VA: {
        name: 'Virginia',
        code: State.VA,
        country: Country.USA,
        number: 78,
    },
    WA: {
        name: 'Washington',
        code: State.WA,
        country: Country.USA,
        number: 53,
    },
    WV: {
        name: 'West Virginia',
        code: State.WV,
        country: Country.USA,
        number: 54,
    },
    WI: {
        name: 'Wisconsin',
        code: State.WI,
        country: Country.USA,
        number: 55,
    },
    WY: {
        name: 'Wyoming',
        code: State.WY,
        country: Country.USA,
        number: 56,
    },
    // Military Base "State" Codes
    // https://support.shippingeasy.com/hc/en-us/articles/203085299-How-to-Format-Military-mail-addresses
    AE: {
        name: 'Armed Forces',
        code: State.AE,
        country: Country.USA,
        number: -1,
    },
    AP: {
        name: 'Armed Forces Pacific',
        code: State.AP,
        country: Country.USA,
        number: -1,
    },
    AA: {
        name: 'Armed Forces America',
        code: State.AA,
        country: Country.USA,
        number: -1,
    },
    // United Kingdom Countries
    // https://en.wikipedia.org/wiki/ISO_3166-2:GB#Countries_and_province
    ENG: {
        name: 'England',
        code: State.ENG,
        country: Country.GBR,
        number: -1,
    },
    NIR: {
        name: 'Northern Ireland',
        code: State.NIR,
        country: Country.GBR,
        number: -1,
    },
    SCT: {
        name: 'Scotland',
        code: State.SCT,
        country: Country.GBR,
        number: -1,
    },
    WLS: {
        name: 'Wales',
        code: State.WLS,
        country: Country.GBR,
        number: -1,
    },
    // Canada States
    // https://pe.usps.com/text/pub28/28apa_005.htm
    AB: {
        name: 'Alberta',
        code: State.AB,
        country: Country.CAN,
        number: -1,
    },
    BC: {
        name: 'British Columbia',
        code: State.BC,
        country: Country.CAN,
        number: -1,
    },
    MB: {
        name: 'Manitoba',
        code: State.MB,
        country: Country.CAN,
        number: -1,
    },
    NB: {
        name: 'New Brunswick',
        code: State.NB,
        country: Country.CAN,
        number: -1,
    },
    NL: {
        name: 'Newfoundland and Labrador',
        code: State.NL,
        country: Country.CAN,
        number: -1,
    },
    NT: {
        name: 'Northwest Territories',
        code: State.NT,
        country: Country.CAN,
        number: -1,
    },
    NS: {
        name: 'Nova Scotia',
        code: State.NS,
        country: Country.CAN,
        number: -1,
    },
    NU: {
        name: 'Nunavat Territory',
        code: State.NU,
        country: Country.CAN,
        number: -1,
    },
    ON: {
        name: 'Ontario',
        code: State.ON,
        country: Country.CAN,
        number: -1,
    },
    PE: {
        name: 'Prince Edward Island',
        code: State.PE,
        country: Country.CAN,
        number: -1,
    },
    QC: {
        name: 'Quebec',
        code: State.QC,
        country: Country.CAN,
        number: -1,
    },
    SK: {
        name: 'Saskatchewan',
        code: State.SK,
        country: Country.CAN,
        number: -1,
    },
    YT: {
        name: 'Yukon Territory',
        code: State.YT,
        country: Country.CAN,
        number: -1,
    },
};
export function getStateDesc(state) {
    const desc = STATE[state];
    return desc ? { ...desc } : null;
}
const aliases = {};
for (const [key, desc] of Object.entries(STATE)) {
    const arr = aliases[key] = aliases[key] || [];
    arr.push(desc.name);
    desc.number > -1 && arr.push(desc.number);
}
export default aliases;
