// @universe/models/esm/src/models/Address.js
import { isNumber, isNumberOrNull, isString, isStringOrNull, } from '@universe/util';
import { Address, isAccuracyTypeOrNull, isCommercialUseTypeOrNull, isCountryOrNull, isDirectionalOrNull, isFacilityTypeOrNull, isPersonalIdentifierOrNull, isResidenceUseTypeOrNull, isStateOrNull, isStreetTypeOrNull, isUnitAbbrOrNull, } from '../types/index.js';
const addressValidator = {
    id: isString,
    createdAt: isNumber,
    updatedAt: isNumber,
    deletedAt: isNumberOrNull,
    pinType: isPersonalIdentifierOrNull,
    pinNum: isStringOrNull,
    care: isStringOrNull,
    facility: isStringOrNull,
    facilityType: isFacilityTypeOrNull,
    number: isStringOrNull,
    streetType: isStreetTypeOrNull,
    streetPreDir: isDirectionalOrNull,
    streetName: isString,
    streetPostDir: isDirectionalOrNull,
    city: isStringOrNull,
    state: isStateOrNull,
    zip: isStringOrNull,
    zip4: isStringOrNull,
    unitAbbr: isUnitAbbrOrNull,
    unitNum: isStringOrNull,
    country: isCountryOrNull,
    owner: isStringOrNull,
    value: isNumberOrNull,
    area: isNumberOrNull,
    // Building Info
    yearBuilt: isNumberOrNull,
    residentialCode: isResidenceUseTypeOrNull,
    commercialCode: isCommercialUseTypeOrNull,
    assessorCode: isStringOrNull,
    footprint: isNumberOrNull,
    stories: isNumberOrNull,
    doors: isNumberOrNull,
    units: isNumberOrNull,
    // Property Info
    squareFeet: isNumberOrNull,
    floors: isNumberOrNull,
    rooms: isNumberOrNull,
    bedrooms: isNumberOrNull,
    bathrooms: isNumberOrNull,
    verifiedAt: isNumberOrNull,
    verification: isStringOrNull,
    carrierRoute: isStringOrNull,
    lineOfTravel: isNumberOrNull,
    deliveryPoint: isNumberOrNull,
    lat: isNumberOrNull,
    lng: isNumberOrNull,
    elv: isNumberOrNull,
    accuracyScore: isNumberOrNull,
    accuracyType: isAccuracyTypeOrNull,
    legLat: isNumberOrNull,
    legLng: isNumberOrNull,
    residents: (v) => Array.isArray(v),
};
export function isAddress(o, log = false) {
    if (!o || typeof o !== 'object') {
        return false;
    }
    let ret = true;
    const addr = o;
    for (const key of Address.keys()) {
        if (!addressValidator[key]?.(addr[key])) {
            log && log(`Unsanitary Data for person "${addr.id}". Invalid value for ${key}: ${JSON.stringify(addr[key], null, 2)}`);
            ret = false;
        }
    }
    return ret;
}
export function isAddressOrNull(o) {
    if (o === null) {
        return true;
    }
    return isAddress(o);
}
