// @universe/models/esm/src/models/Ethnicity2.js
// https://www.ncbi.nlm.nih.gov/books/NBK219753/
// https://www.cdc.gov/nchs/data/dvs/Race_Ethnicity_CodeSet.pdf
// https://www.ncbi.nlm.nih.gov/books/NBK219757/table/ttt00046/?report=objectonly
// https://www2.census.gov/programs-surveys/decennial/2000/phc/phc-t-43/tab01.pdf?#
export const Ethnicity = {
    Asian: [
        {
            name: 'Indian',
            cdc: '2029-7',
            census: 'X',
        },
        {
            name: 'Kashmir',
            census: 'X',
        },
        {
            name: 'Bhutanese',
            cdc: '2031-3',
            census: 'X',
        },
        {
            name: 'Chinese',
            cdc: '2034-7',
            census: 'X',
        },
        {
            name: 'Manchurian',
            census: 'X',
        },
        {
            name: 'Taiwanese',
            cdc: '2035-4',
            census: 'X',
        },
        {
            name: 'Japanese',
            cdc: '2039-6',
            census: 'X',
        },
        {
            name: 'Okinawan',
            cdc: '2043-8',
            census: 'X',
        },
        {
            name: 'Korean',
            cdc: '2040-4',
            census: 'X',
        },
        {
            name: 'Vietnamese',
            cdc: '2047-9',
            census: 'X',
        },
        {
            name: 'Hmong',
            cdc: '2037-0',
            census: 'X',
        },
        {
            name: 'Maldivian',
            cdc: '2049-5',
            census: 'X',
        },
        {
            name: 'Singaporean',
            cdc: '2051-1',
            census: 'X',
        },
        {
            name: 'Punjab',
            census: 'X',
        },
        {
            name: 'Cantonese',
            census: 'X',
        },
        {
            name: 'Mongolian',
            census: 'X',
        },
        {
            name: 'Tibetan',
            census: 'X',
        },
        {
            name: 'Hong Kong',
            census: 'X',
        },
        {
            name: 'Macao',
            census: 'X',
        },
        {
            name: 'Montagnard',
            census: 'X',
        },
        {
            name: 'Bangladeshi',
            cdc: '2030-5',
            census: 'X',
        },
        {
            name: 'Nepali',
            cdc: '2050-3',
            census: 'X',
        },
        {
            name: 'Bengali',
            census: 'X',
        },
        {
            name: 'Pakistani',
            cdc: '2044-6',
            census: 'X',
        },
        {
            name: 'Sri Lankan',
            cdc: '2045-3',
            census: 'X',
        },
        {
            name: 'Sinhalese',
            census: 'X',
        },
        {
            name: 'Veddah',
            census: 'X',
        },
        {
            name: 'Burmese',
            cdc: '2032-1',
            census: 'X',
        },
        {
            name: 'Shan',
            census: 'X',
        },
        {
            name: 'Cambodian',
            cdc: '2033-9',
            census: 'X',
        },
        {
            name: 'Khmer',
            census: 'X',
        },
        {
            name: 'Filipino',
            cdc: '2036-2',
            census: 'X',
        },
        {
            name: 'Indonesian',
            cdc: '2038-8',
            census: 'X',
        },
        {
            name: 'Bornean',
            census: 'X',
        },
        {
            name: 'Javanese/Jawa',
            census: 'X',
        },
        {
            name: 'Sumatra',
            census: 'X',
        },
        {
            name: 'Laotian/Lao',
            cdc: '2041-2',
            census: 'X',
        },
        {
            name: 'Meo',
            census: 'X',
        },
        {
            name: 'Malaysian',
            cdc: '2042-0',
            census: 'X',
        },
        {
            name: 'Thai',
            cdc: '2046-1',
            census: 'X',
        },
    ],
    'Black or African American': [
        {
            name: 'Black',
            cdc: '2056-0',
            census: 'X',
        },
        {
            name: 'African American',
            cdc: '2058-6',
            census: 'X',
        },
        {
            name: 'African',
            cdc: '2060-2',
            census: 'X',
        },
        {
            name: 'Botswanan',
            cdc: '2061-0',
            census: 'X',
        },
        {
            name: 'Djibouti',
            census: 'X',
        },
        {
            name: 'Liberian',
            cdc: '2063-6',
            census: 'X',
        },
        {
            name: 'Namibian',
            cdc: '2064-4',
            census: 'X',
        },
        {
            name: 'Nigerian',
            cdc: '2065-1',
            census: 'X',
        },
        {
            name: 'Fulani/Hausa',
            census: 'X',
        },
        {
            name: 'Ibo/Igbo',
            census: 'X',
        },
        {
            name: 'Tiv',
            census: 'X',
        },
        {
            name: 'Voruba',
            census: 'X',
        },
        {
            name: 'Zairean',
            cdc: '2066-9',
            census: 'X',
        },
        {
            name: 'Angolese',
            census: 'X',
        },
        {
            name: 'Benin',
            census: 'X',
        },
        {
            name: 'Burundi',
            census: 'X',
        },
        {
            name: 'Cameroonian',
            census: 'X',
        },
        {
            name: 'Central African Republic',
            census: 'X',
        },
        {
            name: 'Congolese',
            census: 'X',
        },
        {
            name: 'Ivoirian/Cote d’Ivoire',
            census: 'X',
        },
        {
            name: 'Equatorial Guinea',
            census: 'X',
        },
        {
            name: 'Eritrean',
            census: 'X',
        },
        {
            name: 'Gabon',
            census: 'X',
        },
        {
            name: 'Gambian',
            census: 'X',
        },
        {
            name: 'Ghanaian/Ghanian',
            census: 'X',
        },
        {
            name: 'Guinea',
            census: 'X',
        },
        {
            name: 'Guinea-Bissau',
            census: 'X',
        },
        {
            name: 'Lesotho',
            census: 'X',
        },
        {
            name: 'Libyan',
            census: 'X',
        },
        {
            name: 'Malawi',
            census: 'X',
        },
        {
            name: 'Mali',
            census: 'X',
        },
        {
            name: 'Mauritanian',
            census: 'X',
        },
        {
            name: 'Mauritius',
            census: 'X',
        },
        {
            name: 'Mozambique',
            census: 'X',
        },
        {
            name: 'Rwanda',
            census: 'X',
        },
        {
            name: 'Senegalese/Senegal',
            census: 'X',
        },
        {
            name: 'Afrikaner',
            census: 'X',
        },
        {
            name: 'Natalian',
            census: 'X',
        },
        {
            name: 'Swazi',
            census: 'X',
        },
        {
            name: 'Tanzanian',
            census: 'X',
        },
        {
            name: 'Togolese',
            census: 'X',
        },
        {
            name: 'Ugandan',
            census: 'X',
        },
        {
            name: 'Zambian',
            census: 'X',
        },
        {
            name: 'Zimbabwean',
            census: 'X',
        },
        {
            name: 'Congo Brazzaville',
            census: 'X',
        },
        {
            name: 'Corsico Islander',
            census: 'X',
        },
        {
            name: 'Niger',
            census: 'X',
        },
        {
            name: 'Zulu',
            census: 'X',
        },
        {
            name: 'Tanganyikan',
            census: 'X',
        },
        {
            name: 'Zanzibar Islander',
            census: 'X',
        },
        {
            name: 'Upper Voltan',
            census: 'X',
        },
        {
            name: 'Volta',
            census: 'X',
        },
        {
            name: 'African Islander',
            census: 'X',
        },
        {
            name: 'Central African',
            census: 'X',
        },
        {
            name: 'Eastern African',
            census: 'X',
        },
        {
            name: 'Western African',
            census: 'X',
        },
        {
            name: 'Antiguan/Barbadian/Bajan',
            cdc: '2068-5',
            census: '325',
        },
        {
            name: 'Dominica Islander',
            cdc: '2070-1',
            census: 'X',
        },
        {
            name: 'Bermudan',
            census: 'X',
        },
        {
            name: 'Cayman Islander',
            census: 'X',
        },
        {
            name: 'U.S. Virgin Islander',
            census: 'X',
        },
        {
            name: 'St Croix Islander',
            census: 'X',
        },
        {
            name: 'St John Islander',
            census: 'X',
        },
        {
            name: 'St Thomas Islander',
            census: 'X',
        },
        {
            name: 'French West Indies',
            census: 'X',
        },
        {
            name: 'Guadeloupe Islander',
            census: 'X',
        },
        {
            name: 'Cayenne',
            census: 'X',
        },
        {
            name: 'Nevis Islander/Kittitian/Nevisian',
            census: 'X',
        },
        {
            name: 'Grenadian',
            census: 'X',
        },
        {
            name: 'Vincent-Grenadine Islander',
            census: 'X',
        },
        {
            name: 'St Lucia Islander/Saint Lucian',
            census: 'X',
        },
        {
            name: 'Bahamian',
            cdc: '2067-7',
            census: 'X',
        },
        {
            name: 'Jamaican',
            cdc: '2072-7',
            census: 'X',
        },
        {
            name: 'Haitian',
            cdc: '2071-9',
            census: 'X',
        },
        {
            name: 'Ethiopian',
            cdc: '2062-8',
            census: 'X',
        },
        {
            name: 'Kenyan',
            census: 'X',
        },
        {
            name: 'Sierra Leonean',
            census: 'X',
        },
        {
            name: 'Somali',
            census: 'X',
        },
    ],
    'Native Hawaiian or Other Pacific Islander': [
        {
            name: 'Pacific Islander',
            cdc: '2500-7',
            census: 'X',
        },
        {
            name: 'Native Hawaiian/Hawaiian',
            cdc: '2079-2',
            census: 'X',
        },
        {
            name: 'Polynesian',
            cdc: '2078-4',
            census: 'X',
        },
        {
            name: 'Samoan',
            cdc: '2080-0',
            census: 'X',
        },
        {
            name: 'Tahitian',
            cdc: '2081-8',
            census: 'X',
        },
        {
            name: 'Tongan',
            cdc: '2082-6',
            census: 'X',
        },
        {
            name: 'Tokelauan',
            cdc: '2083-4',
            census: 'X',
        },
        {
            name: 'Micronesian',
            cdc: '2085-9',
            census: 'X',
        },
        {
            name: 'Guamanian',
            cdc: '2087-5',
            census: 'X',
        },
        {
            name: 'Chamorro',
            cdc: '2088-3',
            census: 'X',
        },
        {
            name: 'Mariana Islander',
            cdc: '2089-1',
            census: '',
        },
        {
            name: 'Marshallese',
            cdc: '2090-9',
            census: 'X',
        },
        {
            name: 'Palauan',
            cdc: '2091-7',
            census: 'X',
        },
        {
            name: 'Kosraean',
            cdc: '2093-3',
            census: 'X',
        },
        {
            name: 'Pohnpeian/Ponapean',
            cdc: '2094-1',
            census: 'X',
        },
        {
            name: 'Saipanese',
            cdc: '2095-8',
            census: 'X',
        },
        {
            name: 'Kiribati/Kirabatese',
            cdc: '2096-6',
            census: 'X',
        },
        {
            name: 'Chuukese/Trukese',
            cdc: '2097-4',
            census: 'X',
        },
        {
            name: 'Yapese',
            cdc: '2098-2',
            census: 'X',
        },
        {
            name: 'Melanesian',
            cdc: '2100-6',
            census: 'X',
        },
        {
            name: 'Fijian',
            cdc: '2101-4',
            census: 'X',
        },
        {
            name: 'Papua New Guinean',
            cdc: '2102-2',
            census: 'X',
        },
        {
            name: 'Solomon Islander',
            cdc: '2103-0',
            census: 'X',
        },
        {
            name: 'Vanuatuan',
            cdc: '2104-8',
            census: 'X',
        },
        {
            name: 'Australian',
            census: 'X',
        },
        {
            name: 'Tasmanian',
            census: 'X',
        },
        {
            name: 'Australian Aborigine',
            census: 'X',
        },
        {
            name: 'New Zealander',
            census: 'X',
        },
        {
            name: 'Maori',
            census: 'X',
        },
        {
            name: 'Cook Islander',
            census: 'X',
        },
        {
            name: 'Niuean',
            census: 'X',
        },
        {
            name: 'Nauruan',
            census: 'X',
        },
        {
            name: 'New Caledonian',
            census: 'X',
        },
    ],
    White: [
        {
            name: 'European',
            cdc: '2108-9',
            census: 'X',
        },
        {
            name: 'Southern European',
            census: 'X',
        },
        {
            name: 'Eastern European',
            census: 'X',
        },
        {
            name: 'Central European',
            census: 'X',
        },
        {
            name: 'Western European',
            census: 'X',
        },
        // British
        {
            name: 'British',
            census: 'X',
        },
        {
            name: 'English',
            cdc: '2110-5',
            census: 'X',
        },
        {
            name: 'Irish',
            cdc: '2113-9',
            census: 'X',
        },
        {
            name: 'Scottish',
            cdc: '2116-2',
            census: 'X',
        },
        {
            name: 'Cornish',
            census: 'X',
        },
        {
            name: 'French',
            cdc: '2111-3',
            census: 'X',
        },
        {
            name: 'Breton',
            census: 'X',
        },
        {
            name: 'Occitan',
            census: 'X',
        },
        {
            name: 'German',
            cdc: '2112-1',
            census: 'X',
        },
        {
            name: 'Northern European',
            census: 'X',
        },
        // Scandinavian
        {
            name: 'Scandinavian',
            census: 'X',
        },
        {
            name: 'Swedish',
            census: 'X',
        },
        {
            name: 'Norwegian',
            census: 'X',
        },
        {
            name: 'Finnish',
            census: 'X',
        },
        {
            name: 'Danish',
            census: 'X',
        },
        {
            name: 'Icelander',
            census: 'X',
        },
        {
            name: 'Karelian',
            census: 'X',
        },
        {
            name: 'Faeroe Islander',
            census: 'X',
        },
        {
            name: 'Frisian',
            census: 'X',
        },
        // Dutch
        {
            name: 'Dutch',
            census: 'X',
        },
        {
            name: 'Belgian',
            census: 'X',
        },
        // North American
        {
            name: 'Appalachian',
            census: 'X',
        },
        {
            name: 'Pennsylvania German',
            census: 'X',
        },
        {
            name: 'Canadian',
            census: 'X',
        },
        {
            name: 'French Canadian',
            census: 'X',
        },
        {
            name: 'Portuguese',
            census: 'X',
        },
        {
            name: 'Acadian/Cajun',
            census: 'X',
        },
        {
            name: 'Italian',
            cdc: '2114-7',
            census: 'X',
        },
        {
            name: 'Polish/Pole',
            cdc: '2115-4',
            census: 'X',
        },
        {
            name: 'Greek',
            census: 'X',
        },
        {
            name: 'Basque',
            cdc: '2146-9',
            census: 'X',
        },
        {
            name: 'Azores Islander',
            census: 'X',
        },
        {
            name: 'Madeira Islander',
            census: 'X',
        },
        {
            name: 'Alsatian',
            census: 'X',
        },
        {
            name: 'Andorran',
            census: 'X',
        },
        {
            name: 'Austrian',
            census: 'X',
        },
        {
            name: 'Tirol',
            census: 'X',
        },
        {
            name: 'Walloon',
            census: 'X',
        },
        {
            name: 'Flemish/Fleming',
            census: 'X',
        },
        {
            name: 'Corsican',
            census: 'X',
        },
        {
            name: 'Cypriot',
            census: 'X',
        },
        {
            name: 'Greek Cypriot',
            census: 'X',
        },
        {
            name: 'Turkish Cypriot',
            census: 'X',
        },
        {
            name: 'Lapp',
            census: 'X',
        },
        {
            name: 'Liechtensteiner',
            census: 'X',
        },
        {
            name: 'Luxemburger',
            census: 'X',
        },
        {
            name: 'Maltese',
            census: 'X',
        },
        {
            name: 'Manx',
            census: 'X',
        },
        {
            name: 'Monegasque',
            census: 'X',
        },
        {
            name: 'Aland Islander',
            census: 'X',
        },
        {
            name: 'Swiss',
            census: 'X',
        },
        {
            name: 'Welsh',
            census: 'X',
        },
        {
            name: 'Celtic',
            census: 'X',
        },
        {
            name: 'Bukovina',
            census: 'X',
        },
        {
            name: 'Silesian',
            census: 'X',
        },
        {
            name: 'Galician',
            census: 'X',
        },
        {
            name: 'Carpatho Rusyn',
            census: 'X',
        },
        {
            name: 'Carpathian',
            census: 'X',
        },
        {
            name: 'Rusyn',
            census: 'X',
        },
        {
            name: 'Albanian',
            census: 'X',
        },
        {
            name: 'Azerbaijani',
            census: 'X',
        },
        {
            name: 'Belorussian/Belarusian',
            census: 'X',
        },
        {
            name: 'Bosniak',
            census: '',
        },
        {
            name: 'Bulgarian',
            census: 'X',
        },
        {
            name: 'Croatian',
            census: 'X',
        },
        {
            name: 'Czech',
            census: 'X',
        },
        {
            name: 'Moravian',
            census: 'X',
        },
        {
            name: 'Czechoslovakian',
            census: 'X',
        },
        {
            name: 'Georgian',
            census: 'X',
        },
        {
            name: 'Livonian',
            census: 'X',
        },
        {
            name: 'Hungarian',
            census: '',
        },
        {
            name: 'Magyar',
            census: 'X',
        },
        {
            name: 'Estonian',
            census: 'X',
        },
        {
            name: 'Latvian',
            census: 'X',
        },
        {
            name: 'Lithuanian',
            census: 'X',
        },
        {
            name: 'Macedonian',
            census: 'X',
        },
        {
            name: 'Montenegrin',
            census: 'X',
        },
        {
            name: 'Romanian',
            census: 'X',
        },
        {
            name: 'Moldavian',
            census: 'X',
        },
        {
            name: 'Bessarabian',
            census: 'X',
        },
        {
            name: 'Wallachian',
            census: 'X',
        },
        {
            name: 'Serbian',
            census: 'X',
        },
        {
            name: 'Slovak',
            census: 'X',
        },
        {
            name: 'Slovene',
            census: 'X',
        },
        {
            name: 'Sorbian/Wend',
            census: 'X',
        },
        {
            name: 'Ukrainian',
            census: 'X',
        },
        {
            name: 'Lemko',
            census: 'X',
        },
        {
            name: 'Bioko',
            census: 'X',
        },
        {
            name: 'Husel',
            census: 'X',
        },
        {
            name: 'Russian',
            census: 'X',
        },
        {
            name: 'Muscovite',
            census: 'X',
        },
        {
            name: 'Ruthenian',
            census: 'X',
        },
        {
            name: 'Cossack',
            census: 'X',
        },
        {
            name: 'Bohemian',
            census: 'X',
        },
        {
            name: 'Finno Ugrian',
            census: 'X',
        },
        {
            name: 'Mordovian',
            census: 'X',
        },
        {
            name: 'Voytak',
            census: 'X',
        },
        {
            name: 'Gruziia',
            census: 'X',
        },
        {
            name: 'Soviet Turkic',
            census: 'X',
        },
        {
            name: 'Volga',
            census: 'X',
        },
        {
            name: 'Kalmyk',
            census: 'X',
        },
        {
            name: 'North Caucasian',
            census: 'X',
        },
        {
            name: 'North Caucasian Turkic',
            census: 'X',
        },
        {
            name: 'Ossetian',
            census: 'X',
        },
        {
            name: 'Bashkir',
            census: 'X',
        },
        {
            name: 'Chuvash',
            census: 'X',
        },
        {
            name: 'Gagauz',
            census: 'X',
        },
        {
            name: 'Mesknetian',
            census: 'X',
        },
        {
            name: 'Tuvinian',
            census: 'X',
        },
        {
            name: 'Yakut',
            census: 'X',
        },
        {
            name: 'Tatar',
            census: 'X',
        },
        {
            name: 'Turkestani',
            census: 'X',
        },
        {
            name: 'Uzbekistani',
            census: 'X',
        },
        {
            name: 'Lebanese',
            cdc: '2123-8',
            census: 'X',
        },
        {
            name: 'Windish',
            census: 'X',
        },
        {
            name: 'Yugoslavian',
            census: 'X',
        },
        {
            name: 'Slavic',
            census: 'X',
        },
        {
            name: 'Tajik',
            census: 'X',
        },
        {
            name: 'Slavonian',
            census: 'X',
        },
    ],
    'Middle Eastern or North African': [
        {
            name: 'Syrian',
            cdc: '2125-3',
            census: 'X',
        },
        {
            name: 'Pashtun',
            census: 'X',
        },
        {
            name: 'Israeli',
            cdc: '2127-9',
            census: 'X',
        },
        {
            name: 'Algerian',
            census: 'X',
        },
        {
            name: 'Kurdish',
            census: 'X',
        },
        {
            name: 'North African',
            census: 'X',
        },
        {
            name: 'Berber/Amazigh/Imazighen',
            census: 'X',
        },
        {
            name: 'Bahraini',
            census: 'X',
        },
        {
            name: 'Kuwaiti',
            census: 'X',
        },
        {
            name: 'Omani',
            census: 'X',
        },
        {
            name: 'Qatar',
            census: 'X',
        },
        {
            name: 'Bedouin',
            census: 'X',
        },
        {
            name: 'Emirati',
            census: 'X',
        },
        {
            name: 'Afghani',
            cdc: '2126-1',
            census: 'X',
        },
        {
            name: 'Egyptian',
            cdc: '2120-4',
            census: 'X',
        },
        {
            name: 'Iranian',
            cdc: '2121-2',
            census: 'X',
        },
        {
            name: 'Iraqi',
            cdc: '2122-0',
            census: 'X',
        },
        {
            name: 'Jordanian',
            census: 'X',
        },
        {
            name: 'Armenian',
            cdc: '2109-7',
            census: 'X',
        },
        {
            name: 'Turkish',
            census: 'X',
        },
        {
            name: 'Yemeni',
            census: 'X',
        },
        {
            name: 'Palestinian',
            cdc: '2124-6',
            census: 'X',
        },
        {
            name: 'Assyrian',
            cdc: '2119-6',
            census: 'X',
        },
        {
            name: 'Arabic',
            cdc: '2129-5',
            census: 'X',
        },
    ],
    'American Indian or Alaska Native': [
        {
            name: 'Native American',
            cdc: '1004-1',
            census: '920',
        },
        {
            name: 'Alaska Native',
            cdc: '1735-0',
            census: '',
        },
    ],
    'Hispanic or Latino': [
        {
            name: 'Latino',
            census: '252',
        },
        {
            name: 'Hispanic',
            census: '290',
        },
        {
            name: 'Spanish',
            cdc: '2137-8',
            census: '291',
        },
        {
            name: 'Salvadoran',
            cdc: '2161-8',
            census: 'X',
        },
        {
            name: 'Central American',
            cdc: '2155-0',
            census: 'X',
        },
        {
            name: 'Costa Rican',
            cdc: '2156-8',
            census: 'X',
        },
        {
            name: 'Nicaraguan',
            cdc: '2159-2',
            census: 'X',
        },
        {
            name: 'Panamanian',
            cdc: '2160-0',
            census: 'X',
        },
        {
            name: 'Native Central American',
            cdc: '2162-6',
            census: 'X',
        },
        {
            name: 'Guatemalan',
            cdc: '2157-6',
            census: 'X',
        },
        {
            name: 'Honduran',
            cdc: '2158-4',
            census: 'X',
        },
        {
            name: 'Mexican',
            cdc: '2148-5',
            census: 'X',
        },
        {
            name: 'La Raza',
            cdc: '2152-7',
            census: 'X',
        },
        {
            name: 'Mexican American',
            cdc: '2149-3',
            census: 'X',
        },
        {
            name: 'Mexican Native American',
            cdc: '2153-5',
            census: 'X',
        },
        {
            name: 'South American',
            cdc: '2165-9',
            census: 'X',
        },
        {
            name: 'Argentinean/Argentine',
            cdc: '2166-7',
            census: 'X',
        },
        {
            name: 'Bolivian',
            cdc: '2167-5',
            census: 'X',
        },
        {
            name: 'Chilean',
            cdc: '2168-3',
            census: 'X',
        },
        {
            name: 'Colombian',
            cdc: '2169-1',
            census: 'X',
        },
        {
            name: 'Ecuadorian',
            cdc: '2170-9',
            census: 'X',
        },
        {
            name: 'Paraguayan',
            cdc: '2171-7',
            census: 'X',
        },
        {
            name: 'Peruvian',
            cdc: '2172-5',
            census: 'X',
        },
        {
            name: 'Uruguayan',
            cdc: '2173-3',
            census: 'X',
        },
        {
            name: 'Venezuelan',
            cdc: '2174-1',
            census: 'X',
        },
        {
            name: 'Native South American',
            cdc: '2175-8',
            census: 'X',
        },
        {
            name: 'Puerto Rican',
            cdc: '2180-8',
            census: 'X',
        },
        {
            name: 'Cuban',
            cdc: '2182-4',
            census: 'X',
        },
    ],
};
/*
{
  "name": "Brazilian",
  "census": "X",
},
{
  "name": "Cape Verdean",
  "census": "X",
},
{
  "name": "Belizean",
  "census": "X",
},
{
  "name": "Guyanese",
  "census": "X",
},
{
  "name": "Madagascan",
  "cdc": "2052-9",
  "census": "X",
},
{
  "name": "Tunisian",
  "census": "X",
},
{
  "name": "Surinam",
  "census": "X",
},
{
  "name": "Trinidadian/Tobagoan",
  "cdc": "2074-3",
  "census": "X",
},
{
  "name": "West Indian",
  "cdc": "2075-0",
  "census": "X",
},
{
  "name": "British Virgin Islander",
  "census": "X",
},
{
  "name": "British West Indian",
  "census": "X",
},
{
  "name": "Moroccan",
  "census": "X",
},
{
  "name": "Dominican",
  "cdc": "2069-3",
  "census": "X",
},
{
  "name": "South African",
  "census": "X",
},
{
  "name": "Sudanese",
  "census": "X",
},
{
  "name": "Dinka",
  "census": "X",
},
{
  "name": "Nuer",
  "census": "X",
},
{
  "name": "Fur",
  "census": "X",
},
{
  "name": "Baggara",
  "census": "X",
},
{
  "name": "East Indian",
  "census": "X",
},
*/
