// @universe/models/esm/src/types/diffs.js
import * as helpers from './helpers.js';
export var ModelType;
(function (ModelType) {
    ModelType["METADATA"] = "Metadata";
    ModelType["ADDRESS"] = "Address";
    ModelType["TURF"] = "Turf";
    ModelType["EMAIL"] = "Email";
    ModelType["PHONE"] = "Phone";
    ModelType["PERSON_ADDRESS"] = "PersonAddress";
    ModelType["PERSON_PHONE"] = "PersonPhone";
    ModelType["ACCOUNT"] = "Account";
    ModelType["EXPERIENCE"] = "Experience";
    ModelType["CAMPAIGN"] = "Campaign";
    ModelType["CONTEST"] = "Contest";
    ModelType["ELECTION"] = "Election";
    ModelType["VOTE"] = "Vote";
    ModelType["RETURNS"] = "Returns";
    ModelType["SCORE"] = "Score";
    ModelType["PERSON"] = "Person";
})(ModelType = ModelType || (ModelType = {}));
export function isModelType(name) {
    return name === ModelType.METADATA || name === ModelType.ADDRESS || name === ModelType.TURF || name === ModelType.EMAIL || name === ModelType.PHONE || name === ModelType.PERSON_ADDRESS || name === ModelType.PERSON_PHONE || name === ModelType.ACCOUNT || name === ModelType.EXPERIENCE || name === ModelType.CAMPAIGN || name === ModelType.CONTEST || name === ModelType.ELECTION || name === ModelType.VOTE || name === ModelType.RETURNS || name === ModelType.SCORE || name === ModelType.PERSON || false;
}
export const ModelKeysLookup = {
    MetadataKeys: helpers.Metadata.keys(),
    AddressKeys: helpers.Address.keys(),
    TurfKeys: helpers.Turf.keys(),
    EmailKeys: helpers.Email.keys(),
    PhoneKeys: helpers.Phone.keys(),
    PersonAddressKeys: helpers.PersonAddress.keys(),
    PersonPhoneKeys: helpers.PersonPhone.keys(),
    AccountKeys: helpers.Account.keys(),
    ExperienceKeys: helpers.Experience.keys(),
    CampaignKeys: helpers.Campaign.keys(),
    ContestKeys: helpers.Contest.keys(),
    ElectionKeys: helpers.Election.keys(),
    VoteKeys: helpers.Vote.keys(),
    ReturnsKeys: helpers.Returns.keys(),
    ScoreKeys: helpers.Score.keys(),
    PersonKeys: helpers.Person.keys(),
};
export const ModelTypesLookup = {
    MetadataTypes: helpers.Metadata.types(),
    AddressTypes: helpers.Address.types(),
    TurfTypes: helpers.Turf.types(),
    EmailTypes: helpers.Email.types(),
    PhoneTypes: helpers.Phone.types(),
    PersonAddressTypes: helpers.PersonAddress.types(),
    PersonPhoneTypes: helpers.PersonPhone.types(),
    AccountTypes: helpers.Account.types(),
    ExperienceTypes: helpers.Experience.types(),
    CampaignTypes: helpers.Campaign.types(),
    ContestTypes: helpers.Contest.types(),
    ElectionTypes: helpers.Election.types(),
    VoteTypes: helpers.Vote.types(),
    ReturnsTypes: helpers.Returns.types(),
    ScoreTypes: helpers.Score.types(),
    PersonTypes: helpers.Person.types(),
};
export function diffMetadata(before, after) {
    const beforeDiff = before ? ({
        key: before.key,
        value: (before?.value !== undefined && before?.value !== after?.value) ? before?.value : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        key: after.key,
        value: (after?.value !== undefined && before?.value !== after?.value && ++changes) ? after?.value : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.METADATA,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffAddress(before, after) {
    const beforeDiff = before ? ({
        id: before.id,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        care: (before?.care !== undefined && before?.care !== after?.care) ? before?.care : undefined,
        facility: (before?.facility !== undefined && before?.facility !== after?.facility) ? before?.facility : undefined,
        facilityType: (before?.facilityType !== undefined && before?.facilityType !== after?.facilityType) ? before?.facilityType : undefined,
        pinNum: (before?.pinNum !== undefined && before?.pinNum !== after?.pinNum) ? before?.pinNum : undefined,
        pinType: (before?.pinType !== undefined && before?.pinType !== after?.pinType) ? before?.pinType : undefined,
        number: (before?.number !== undefined && before?.number !== after?.number) ? before?.number : undefined,
        streetPreDir: (before?.streetPreDir !== undefined && before?.streetPreDir !== after?.streetPreDir) ? before?.streetPreDir : undefined,
        streetName: (before?.streetName !== undefined && before?.streetName !== after?.streetName) ? before?.streetName : undefined,
        streetType: (before?.streetType !== undefined && before?.streetType !== after?.streetType) ? before?.streetType : undefined,
        streetPostDir: (before?.streetPostDir !== undefined && before?.streetPostDir !== after?.streetPostDir) ? before?.streetPostDir : undefined,
        unitAbbr: (before?.unitAbbr !== undefined && before?.unitAbbr !== after?.unitAbbr) ? before?.unitAbbr : undefined,
        unitNum: (before?.unitNum !== undefined && before?.unitNum !== after?.unitNum) ? before?.unitNum : undefined,
        city: (before?.city !== undefined && before?.city !== after?.city) ? before?.city : undefined,
        state: (before?.state !== undefined && before?.state !== after?.state) ? before?.state : undefined,
        zip: (before?.zip !== undefined && before?.zip !== after?.zip) ? before?.zip : undefined,
        zip4: (before?.zip4 !== undefined && before?.zip4 !== after?.zip4) ? before?.zip4 : undefined,
        country: (before?.country !== undefined && before?.country !== after?.country) ? before?.country : undefined,
        verifiedAt: (before?.verifiedAt !== undefined && before?.verifiedAt !== after?.verifiedAt) ? before?.verifiedAt : undefined,
        verification: (before?.verification !== undefined && before?.verification !== after?.verification) ? before?.verification : undefined,
        carrierRoute: (before?.carrierRoute !== undefined && before?.carrierRoute !== after?.carrierRoute) ? before?.carrierRoute : undefined,
        lineOfTravel: (before?.lineOfTravel !== undefined && before?.lineOfTravel !== after?.lineOfTravel) ? before?.lineOfTravel : undefined,
        deliveryPoint: (before?.deliveryPoint !== undefined && before?.deliveryPoint !== after?.deliveryPoint) ? before?.deliveryPoint : undefined,
        lat: (before?.lat !== undefined && before?.lat !== after?.lat) ? before?.lat : undefined,
        lng: (before?.lng !== undefined && before?.lng !== after?.lng) ? before?.lng : undefined,
        elv: (before?.elv !== undefined && before?.elv !== after?.elv) ? before?.elv : undefined,
        accuracyScore: (before?.accuracyScore !== undefined && before?.accuracyScore !== after?.accuracyScore) ? before?.accuracyScore : undefined,
        accuracyType: (before?.accuracyType !== undefined && before?.accuracyType !== after?.accuracyType) ? before?.accuracyType : undefined,
        legLat: (before?.legLat !== undefined && before?.legLat !== after?.legLat) ? before?.legLat : undefined,
        legLng: (before?.legLng !== undefined && before?.legLng !== after?.legLng) ? before?.legLng : undefined,
        owner: (before?.owner !== undefined && before?.owner !== after?.owner) ? before?.owner : undefined,
        value: (before?.value !== undefined && before?.value !== after?.value) ? before?.value : undefined,
        area: (before?.area !== undefined && before?.area !== after?.area) ? before?.area : undefined,
        yearBuilt: (before?.yearBuilt !== undefined && before?.yearBuilt !== after?.yearBuilt) ? before?.yearBuilt : undefined,
        residentialCode: (before?.residentialCode !== undefined && before?.residentialCode !== after?.residentialCode) ? before?.residentialCode : undefined,
        commercialCode: (before?.commercialCode !== undefined && before?.commercialCode !== after?.commercialCode) ? before?.commercialCode : undefined,
        assessorCode: (before?.assessorCode !== undefined && before?.assessorCode !== after?.assessorCode) ? before?.assessorCode : undefined,
        footprint: (before?.footprint !== undefined && before?.footprint !== after?.footprint) ? before?.footprint : undefined,
        stories: (before?.stories !== undefined && before?.stories !== after?.stories) ? before?.stories : undefined,
        doors: (before?.doors !== undefined && before?.doors !== after?.doors) ? before?.doors : undefined,
        units: (before?.units !== undefined && before?.units !== after?.units) ? before?.units : undefined,
        squareFeet: (before?.squareFeet !== undefined && before?.squareFeet !== after?.squareFeet) ? before?.squareFeet : undefined,
        floors: (before?.floors !== undefined && before?.floors !== after?.floors) ? before?.floors : undefined,
        rooms: (before?.rooms !== undefined && before?.rooms !== after?.rooms) ? before?.rooms : undefined,
        bedrooms: (before?.bedrooms !== undefined && before?.bedrooms !== after?.bedrooms) ? before?.bedrooms : undefined,
        bathrooms: (before?.bathrooms !== undefined && before?.bathrooms !== after?.bathrooms) ? before?.bathrooms : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        id: after.id,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        care: (after?.care !== undefined && before?.care !== after?.care && ++changes) ? after?.care : undefined,
        facility: (after?.facility !== undefined && before?.facility !== after?.facility && ++changes) ? after?.facility : undefined,
        facilityType: (after?.facilityType !== undefined && before?.facilityType !== after?.facilityType && ++changes) ? after?.facilityType : undefined,
        pinNum: (after?.pinNum !== undefined && before?.pinNum !== after?.pinNum && ++changes) ? after?.pinNum : undefined,
        pinType: (after?.pinType !== undefined && before?.pinType !== after?.pinType && ++changes) ? after?.pinType : undefined,
        number: (after?.number !== undefined && before?.number !== after?.number && ++changes) ? after?.number : undefined,
        streetPreDir: (after?.streetPreDir !== undefined && before?.streetPreDir !== after?.streetPreDir && ++changes) ? after?.streetPreDir : undefined,
        streetName: (after?.streetName !== undefined && before?.streetName !== after?.streetName && ++changes) ? after?.streetName : undefined,
        streetType: (after?.streetType !== undefined && before?.streetType !== after?.streetType && ++changes) ? after?.streetType : undefined,
        streetPostDir: (after?.streetPostDir !== undefined && before?.streetPostDir !== after?.streetPostDir && ++changes) ? after?.streetPostDir : undefined,
        unitAbbr: (after?.unitAbbr !== undefined && before?.unitAbbr !== after?.unitAbbr && ++changes) ? after?.unitAbbr : undefined,
        unitNum: (after?.unitNum !== undefined && before?.unitNum !== after?.unitNum && ++changes) ? after?.unitNum : undefined,
        city: (after?.city !== undefined && before?.city !== after?.city && ++changes) ? after?.city : undefined,
        state: (after?.state !== undefined && before?.state !== after?.state && ++changes) ? after?.state : undefined,
        zip: (after?.zip !== undefined && before?.zip !== after?.zip && ++changes) ? after?.zip : undefined,
        zip4: (after?.zip4 !== undefined && before?.zip4 !== after?.zip4 && ++changes) ? after?.zip4 : undefined,
        country: (after?.country !== undefined && before?.country !== after?.country && ++changes) ? after?.country : undefined,
        verifiedAt: (after?.verifiedAt !== undefined && before?.verifiedAt !== after?.verifiedAt && ++changes) ? after?.verifiedAt : undefined,
        verification: (after?.verification !== undefined && before?.verification !== after?.verification && ++changes) ? after?.verification : undefined,
        carrierRoute: (after?.carrierRoute !== undefined && before?.carrierRoute !== after?.carrierRoute && ++changes) ? after?.carrierRoute : undefined,
        lineOfTravel: (after?.lineOfTravel !== undefined && before?.lineOfTravel !== after?.lineOfTravel && ++changes) ? after?.lineOfTravel : undefined,
        deliveryPoint: (after?.deliveryPoint !== undefined && before?.deliveryPoint !== after?.deliveryPoint && ++changes) ? after?.deliveryPoint : undefined,
        lat: (after?.lat !== undefined && before?.lat !== after?.lat && ++changes) ? after?.lat : undefined,
        lng: (after?.lng !== undefined && before?.lng !== after?.lng && ++changes) ? after?.lng : undefined,
        elv: (after?.elv !== undefined && before?.elv !== after?.elv && ++changes) ? after?.elv : undefined,
        accuracyScore: (after?.accuracyScore !== undefined && before?.accuracyScore !== after?.accuracyScore && ++changes) ? after?.accuracyScore : undefined,
        accuracyType: (after?.accuracyType !== undefined && before?.accuracyType !== after?.accuracyType && ++changes) ? after?.accuracyType : undefined,
        legLat: (after?.legLat !== undefined && before?.legLat !== after?.legLat && ++changes) ? after?.legLat : undefined,
        legLng: (after?.legLng !== undefined && before?.legLng !== after?.legLng && ++changes) ? after?.legLng : undefined,
        owner: (after?.owner !== undefined && before?.owner !== after?.owner && ++changes) ? after?.owner : undefined,
        value: (after?.value !== undefined && before?.value !== after?.value && ++changes) ? after?.value : undefined,
        area: (after?.area !== undefined && before?.area !== after?.area && ++changes) ? after?.area : undefined,
        yearBuilt: (after?.yearBuilt !== undefined && before?.yearBuilt !== after?.yearBuilt && ++changes) ? after?.yearBuilt : undefined,
        residentialCode: (after?.residentialCode !== undefined && before?.residentialCode !== after?.residentialCode && ++changes) ? after?.residentialCode : undefined,
        commercialCode: (after?.commercialCode !== undefined && before?.commercialCode !== after?.commercialCode && ++changes) ? after?.commercialCode : undefined,
        assessorCode: (after?.assessorCode !== undefined && before?.assessorCode !== after?.assessorCode && ++changes) ? after?.assessorCode : undefined,
        footprint: (after?.footprint !== undefined && before?.footprint !== after?.footprint && ++changes) ? after?.footprint : undefined,
        stories: (after?.stories !== undefined && before?.stories !== after?.stories && ++changes) ? after?.stories : undefined,
        doors: (after?.doors !== undefined && before?.doors !== after?.doors && ++changes) ? after?.doors : undefined,
        units: (after?.units !== undefined && before?.units !== after?.units && ++changes) ? after?.units : undefined,
        squareFeet: (after?.squareFeet !== undefined && before?.squareFeet !== after?.squareFeet && ++changes) ? after?.squareFeet : undefined,
        floors: (after?.floors !== undefined && before?.floors !== after?.floors && ++changes) ? after?.floors : undefined,
        rooms: (after?.rooms !== undefined && before?.rooms !== after?.rooms && ++changes) ? after?.rooms : undefined,
        bedrooms: (after?.bedrooms !== undefined && before?.bedrooms !== after?.bedrooms && ++changes) ? after?.bedrooms : undefined,
        bathrooms: (after?.bathrooms !== undefined && before?.bathrooms !== after?.bathrooms && ++changes) ? after?.bathrooms : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.ADDRESS,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffTurf(before, after) {
    const beforeDiff = before ? ({
        id: before.id,
        shape: (before?.shape !== undefined && before?.shape !== after?.shape) ? before?.shape : undefined,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        type: (before?.type !== undefined && before?.type !== after?.type) ? before?.type : undefined,
        state: (before?.state !== undefined && before?.state !== after?.state) ? before?.state : undefined,
        county: (before?.county !== undefined && before?.county !== after?.county) ? before?.county : undefined,
        group: (before?.group !== undefined && before?.group !== after?.group) ? before?.group : undefined,
        name: (before?.name !== undefined && before?.name !== after?.name) ? before?.name : undefined,
        lat: (before?.lat !== undefined && before?.lat !== after?.lat) ? before?.lat : undefined,
        lng: (before?.lng !== undefined && before?.lng !== after?.lng) ? before?.lng : undefined,
        latMin: (before?.latMin !== undefined && before?.latMin !== after?.latMin) ? before?.latMin : undefined,
        latMax: (before?.latMax !== undefined && before?.latMax !== after?.latMax) ? before?.latMax : undefined,
        lngMin: (before?.lngMin !== undefined && before?.lngMin !== after?.lngMin) ? before?.lngMin : undefined,
        lngMax: (before?.lngMax !== undefined && before?.lngMax !== after?.lngMax) ? before?.lngMax : undefined,
        area: (before?.area !== undefined && before?.area !== after?.area) ? before?.area : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        id: after.id,
        shape: (after?.shape !== undefined && before?.shape !== after?.shape && ++changes) ? after?.shape : undefined,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        type: (after?.type !== undefined && before?.type !== after?.type && ++changes) ? after?.type : undefined,
        state: (after?.state !== undefined && before?.state !== after?.state && ++changes) ? after?.state : undefined,
        county: (after?.county !== undefined && before?.county !== after?.county && ++changes) ? after?.county : undefined,
        group: (after?.group !== undefined && before?.group !== after?.group && ++changes) ? after?.group : undefined,
        name: (after?.name !== undefined && before?.name !== after?.name && ++changes) ? after?.name : undefined,
        lat: (after?.lat !== undefined && before?.lat !== after?.lat && ++changes) ? after?.lat : undefined,
        lng: (after?.lng !== undefined && before?.lng !== after?.lng && ++changes) ? after?.lng : undefined,
        latMin: (after?.latMin !== undefined && before?.latMin !== after?.latMin && ++changes) ? after?.latMin : undefined,
        latMax: (after?.latMax !== undefined && before?.latMax !== after?.latMax && ++changes) ? after?.latMax : undefined,
        lngMin: (after?.lngMin !== undefined && before?.lngMin !== after?.lngMin && ++changes) ? after?.lngMin : undefined,
        lngMax: (after?.lngMax !== undefined && before?.lngMax !== after?.lngMax && ++changes) ? after?.lngMax : undefined,
        area: (after?.area !== undefined && before?.area !== after?.area && ++changes) ? after?.area : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.TURF,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffEmail(before, after) {
    const beforeDiff = before ? ({
        email: before.email,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        deliverable: (before?.deliverable !== undefined && before?.deliverable !== after?.deliverable) ? before?.deliverable : undefined,
        personId: before.personId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        email: after.email,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        deliverable: (after?.deliverable !== undefined && before?.deliverable !== after?.deliverable && ++changes) ? after?.deliverable : undefined,
        personId: after.personId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.EMAIL,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffPhone(before, after) {
    const beforeDiff = before ? ({
        number: before.number,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        system: (before?.system !== undefined && before?.system !== after?.system) ? before?.system : undefined,
        sms: (before?.sms !== undefined && before?.sms !== after?.sms) ? before?.sms : undefined,
        voice: (before?.voice !== undefined && before?.voice !== after?.voice) ? before?.voice : undefined,
        ndnc: (before?.ndnc !== undefined && before?.ndnc !== after?.ndnc) ? before?.ndnc : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        number: after.number,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        system: (after?.system !== undefined && before?.system !== after?.system && ++changes) ? after?.system : undefined,
        sms: (after?.sms !== undefined && before?.sms !== after?.sms && ++changes) ? after?.sms : undefined,
        voice: (after?.voice !== undefined && before?.voice !== after?.voice && ++changes) ? after?.voice : undefined,
        ndnc: (after?.ndnc !== undefined && before?.ndnc !== after?.ndnc && ++changes) ? after?.ndnc : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.PHONE,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffPersonAddress(before, after) {
    const beforeDiff = before ? ({
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        ncoa: (before?.ncoa !== undefined && before?.ncoa !== after?.ncoa) ? before?.ncoa : undefined,
        type: (before?.type !== undefined && before?.type !== after?.type) ? before?.type : undefined,
        personId: before.personId,
        addressId: before.addressId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        ncoa: (after?.ncoa !== undefined && before?.ncoa !== after?.ncoa && ++changes) ? after?.ncoa : undefined,
        type: (after?.type !== undefined && before?.type !== after?.type && ++changes) ? after?.type : undefined,
        personId: after.personId,
        addressId: after.addressId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.PERSON_ADDRESS,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffPersonPhone(before, after) {
    const beforeDiff = before ? ({
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        type: (before?.type !== undefined && before?.type !== after?.type) ? before?.type : undefined,
        personId: before.personId,
        phoneId: before.phoneId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        type: (after?.type !== undefined && before?.type !== after?.type && ++changes) ? after?.type : undefined,
        personId: after.personId,
        phoneId: after.phoneId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.PERSON_PHONE,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffAccount(before, after) {
    const beforeDiff = before ? ({
        accountType: before.accountType,
        accountId: before.accountId,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        personId: before.personId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        accountType: after.accountType,
        accountId: after.accountId,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        personId: after.personId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.ACCOUNT,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffExperience(before, after) {
    const beforeDiff = before ? ({
        type: before.type,
        employer: before.employer,
        role: before.role,
        startDate: (before?.startDate !== undefined && before?.startDate !== after?.startDate) ? before?.startDate : undefined,
        endDate: (before?.endDate !== undefined && before?.endDate !== after?.endDate) ? before?.endDate : undefined,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        personId: before.personId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        type: after.type,
        employer: after.employer,
        role: after.role,
        startDate: (after?.startDate !== undefined && before?.startDate !== after?.startDate && ++changes) ? after?.startDate : undefined,
        endDate: (after?.endDate !== undefined && before?.endDate !== after?.endDate && ++changes) ? after?.endDate : undefined,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        personId: after.personId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.EXPERIENCE,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffCampaign(before, after) {
    const beforeDiff = before ? ({
        id: before.id,
        name: (before?.name !== undefined && before?.name !== after?.name) ? before?.name : undefined,
        contestId: (before?.contestId !== undefined && before?.contestId !== after?.contestId) ? before?.contestId : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        id: after.id,
        name: (after?.name !== undefined && before?.name !== after?.name && ++changes) ? after?.name : undefined,
        contestId: (after?.contestId !== undefined && before?.contestId !== after?.contestId && ++changes) ? after?.contestId : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.CAMPAIGN,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffContest(before, after) {
    const beforeDiff = before ? ({
        id: before.id,
        name: (before?.name !== undefined && before?.name !== after?.name) ? before?.name : undefined,
        electionId: (before?.electionId !== undefined && before?.electionId !== after?.electionId) ? before?.electionId : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        id: after.id,
        name: (after?.name !== undefined && before?.name !== after?.name && ++changes) ? after?.name : undefined,
        electionId: (after?.electionId !== undefined && before?.electionId !== after?.electionId && ++changes) ? after?.electionId : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.CONTEST,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffElection(before, after) {
    const beforeDiff = before ? ({
        id: before.id,
        electionDate: (before?.electionDate !== undefined && before?.electionDate !== after?.electionDate) ? before?.electionDate : undefined,
        name: (before?.name !== undefined && before?.name !== after?.name) ? before?.name : undefined,
        type: (before?.type !== undefined && before?.type !== after?.type) ? before?.type : undefined,
        turf: (before?.turf !== undefined && before?.turf !== after?.turf) ? before?.turf : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        id: after.id,
        electionDate: (after?.electionDate !== undefined && before?.electionDate !== after?.electionDate && ++changes) ? after?.electionDate : undefined,
        name: (after?.name !== undefined && before?.name !== after?.name && ++changes) ? after?.name : undefined,
        type: (after?.type !== undefined && before?.type !== after?.type && ++changes) ? after?.type : undefined,
        turf: (after?.turf !== undefined && before?.turf !== after?.turf && ++changes) ? after?.turf : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.ELECTION,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffVote(before, after) {
    const beforeDiff = before ? ({
        method: (before?.method !== undefined && before?.method !== after?.method) ? before?.method : undefined,
        party: (before?.party !== undefined && before?.party !== after?.party) ? before?.party : undefined,
        issued: (before?.issued !== undefined && before?.issued !== after?.issued) ? before?.issued : undefined,
        returned: (before?.returned !== undefined && before?.returned !== after?.returned) ? before?.returned : undefined,
        counted: (before?.counted !== undefined && before?.counted !== after?.counted) ? before?.counted : undefined,
        personId: before.personId,
        electionId: before.electionId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        method: (after?.method !== undefined && before?.method !== after?.method && ++changes) ? after?.method : undefined,
        party: (after?.party !== undefined && before?.party !== after?.party && ++changes) ? after?.party : undefined,
        issued: (after?.issued !== undefined && before?.issued !== after?.issued && ++changes) ? after?.issued : undefined,
        returned: (after?.returned !== undefined && before?.returned !== after?.returned && ++changes) ? after?.returned : undefined,
        counted: (after?.counted !== undefined && before?.counted !== after?.counted && ++changes) ? after?.counted : undefined,
        personId: after.personId,
        electionId: after.electionId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.VOTE,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffReturns(before, after) {
    const beforeDiff = before ? ({
        party: before.party,
        rank: before.rank,
        tally: (before?.tally !== undefined && before?.tally !== after?.tally) ? before?.tally : undefined,
        contestId: before.contestId,
        turfId: before.turfId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        party: after.party,
        rank: after.rank,
        tally: (after?.tally !== undefined && before?.tally !== after?.tally && ++changes) ? after?.tally : undefined,
        contestId: after.contestId,
        turfId: after.turfId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.RETURNS,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffScore(before, after) {
    const beforeDiff = before ? ({
        name: before.name,
        value: (before?.value !== undefined && before?.value !== after?.value) ? before?.value : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        personId: before.personId,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        name: after.name,
        value: (after?.value !== undefined && before?.value !== after?.value && ++changes) ? after?.value : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        personId: after.personId,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.SCORE,
        before: beforeDiff,
        after: afterDiff,
    };
}
export function diffPerson(before, after) {
    const beforeDiff = before ? ({
        id: before.id,
        countyId: (before?.countyId !== undefined && before?.countyId !== after?.countyId) ? before?.countyId : undefined,
        affidavitNum: (before?.affidavitNum !== undefined && before?.affidavitNum !== after?.affidavitNum) ? before?.affidavitNum : undefined,
        precinct: (before?.precinct !== undefined && before?.precinct !== after?.precinct) ? before?.precinct : undefined,
        createdAt: (before?.createdAt !== undefined && before?.createdAt !== after?.createdAt) ? before?.createdAt : undefined,
        updatedAt: (before?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt) ? before?.updatedAt : undefined,
        deletedAt: (before?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt) ? before?.deletedAt : undefined,
        preNominal: (before?.preNominal !== undefined && before?.preNominal !== after?.preNominal) ? before?.preNominal : undefined,
        firstName: (before?.firstName !== undefined && before?.firstName !== after?.firstName) ? before?.firstName : undefined,
        middleName: (before?.middleName !== undefined && before?.middleName !== after?.middleName) ? before?.middleName : undefined,
        lastName: (before?.lastName !== undefined && before?.lastName !== after?.lastName) ? before?.lastName : undefined,
        suffix: (before?.suffix !== undefined && before?.suffix !== after?.suffix) ? before?.suffix : undefined,
        postNominal: (before?.postNominal !== undefined && before?.postNominal !== after?.postNominal) ? before?.postNominal : undefined,
        nickname: (before?.nickname !== undefined && before?.nickname !== after?.nickname) ? before?.nickname : undefined,
        birthDate: (before?.birthDate !== undefined && before?.birthDate !== after?.birthDate) ? before?.birthDate : undefined,
        birthState: (before?.birthState !== undefined && before?.birthState !== after?.birthState) ? before?.birthState : undefined,
        birthCountry: (before?.birthCountry !== undefined && before?.birthCountry !== after?.birthCountry) ? before?.birthCountry : undefined,
        gender: (before?.gender !== undefined && before?.gender !== after?.gender) ? before?.gender : undefined,
        party: (before?.party !== undefined && before?.party !== after?.party) ? before?.party : undefined,
        ballot: (before?.ballot !== undefined && before?.ballot !== after?.ballot) ? before?.ballot : undefined,
        religion: (before?.religion !== undefined && before?.religion !== after?.religion) ? before?.religion : undefined,
        education: (before?.education !== undefined && before?.education !== after?.education) ? before?.education : undefined,
        military: (before?.military !== undefined && before?.military !== after?.military) ? before?.military : undefined,
        income: (before?.income !== undefined && before?.income !== after?.income) ? before?.income : undefined,
        primaryLanguageCode: (before?.primaryLanguageCode !== undefined && before?.primaryLanguageCode !== after?.primaryLanguageCode) ? before?.primaryLanguageCode : undefined,
        secondaryLanguageCode: (before?.secondaryLanguageCode !== undefined && before?.secondaryLanguageCode !== after?.secondaryLanguageCode) ? before?.secondaryLanguageCode : undefined,
        primaryEthnicityCode: (before?.primaryEthnicityCode !== undefined && before?.primaryEthnicityCode !== after?.primaryEthnicityCode) ? before?.primaryEthnicityCode : undefined,
        secondaryEthnicityCode: (before?.secondaryEthnicityCode !== undefined && before?.secondaryEthnicityCode !== after?.secondaryEthnicityCode) ? before?.secondaryEthnicityCode : undefined,
    }) : null;
    let changes = 0;
    const afterDiff = after ? ({
        id: after.id,
        countyId: (after?.countyId !== undefined && before?.countyId !== after?.countyId && ++changes) ? after?.countyId : undefined,
        affidavitNum: (after?.affidavitNum !== undefined && before?.affidavitNum !== after?.affidavitNum && ++changes) ? after?.affidavitNum : undefined,
        precinct: (after?.precinct !== undefined && before?.precinct !== after?.precinct && ++changes) ? after?.precinct : undefined,
        createdAt: (after?.createdAt !== undefined && before?.createdAt !== after?.createdAt && ++changes) ? after?.createdAt : undefined,
        updatedAt: (after?.updatedAt !== undefined && before?.updatedAt !== after?.updatedAt && ++changes) ? after?.updatedAt : undefined,
        deletedAt: (after?.deletedAt !== undefined && before?.deletedAt !== after?.deletedAt && ++changes) ? after?.deletedAt : undefined,
        preNominal: (after?.preNominal !== undefined && before?.preNominal !== after?.preNominal && ++changes) ? after?.preNominal : undefined,
        firstName: (after?.firstName !== undefined && before?.firstName !== after?.firstName && ++changes) ? after?.firstName : undefined,
        middleName: (after?.middleName !== undefined && before?.middleName !== after?.middleName && ++changes) ? after?.middleName : undefined,
        lastName: (after?.lastName !== undefined && before?.lastName !== after?.lastName && ++changes) ? after?.lastName : undefined,
        suffix: (after?.suffix !== undefined && before?.suffix !== after?.suffix && ++changes) ? after?.suffix : undefined,
        postNominal: (after?.postNominal !== undefined && before?.postNominal !== after?.postNominal && ++changes) ? after?.postNominal : undefined,
        nickname: (after?.nickname !== undefined && before?.nickname !== after?.nickname && ++changes) ? after?.nickname : undefined,
        birthDate: (after?.birthDate !== undefined && before?.birthDate !== after?.birthDate && ++changes) ? after?.birthDate : undefined,
        birthState: (after?.birthState !== undefined && before?.birthState !== after?.birthState && ++changes) ? after?.birthState : undefined,
        birthCountry: (after?.birthCountry !== undefined && before?.birthCountry !== after?.birthCountry && ++changes) ? after?.birthCountry : undefined,
        gender: (after?.gender !== undefined && before?.gender !== after?.gender && ++changes) ? after?.gender : undefined,
        party: (after?.party !== undefined && before?.party !== after?.party && ++changes) ? after?.party : undefined,
        ballot: (after?.ballot !== undefined && before?.ballot !== after?.ballot && ++changes) ? after?.ballot : undefined,
        religion: (after?.religion !== undefined && before?.religion !== after?.religion && ++changes) ? after?.religion : undefined,
        education: (after?.education !== undefined && before?.education !== after?.education && ++changes) ? after?.education : undefined,
        military: (after?.military !== undefined && before?.military !== after?.military && ++changes) ? after?.military : undefined,
        income: (after?.income !== undefined && before?.income !== after?.income && ++changes) ? after?.income : undefined,
        primaryLanguageCode: (after?.primaryLanguageCode !== undefined && before?.primaryLanguageCode !== after?.primaryLanguageCode && ++changes) ? after?.primaryLanguageCode : undefined,
        secondaryLanguageCode: (after?.secondaryLanguageCode !== undefined && before?.secondaryLanguageCode !== after?.secondaryLanguageCode && ++changes) ? after?.secondaryLanguageCode : undefined,
        primaryEthnicityCode: (after?.primaryEthnicityCode !== undefined && before?.primaryEthnicityCode !== after?.primaryEthnicityCode && ++changes) ? after?.primaryEthnicityCode : undefined,
        secondaryEthnicityCode: (after?.secondaryEthnicityCode !== undefined && before?.secondaryEthnicityCode !== after?.secondaryEthnicityCode && ++changes) ? after?.secondaryEthnicityCode : undefined,
    }) : null;
    if (before && after && changes === 0) {
        return null;
    }
    return {
        model: ModelType.PERSON,
        before: beforeDiff,
        after: afterDiff,
    };
}
